--- src/widget.c
+++ src/widget.c
@@ -1441,5 +1441,5 @@
     while (in->point > 0) {
       wchar_t c;
-      char *p = in->buffer + charpos(in,in->point);
+      char *p = in->buffer + charpos(in,in->point-1);
       size_t res = mbrtowc(&c, p, strlen(p), &mbs);
       if (*p && (res <= 0 || !(iswspace (c) || iswpunct (c))))
@@ -1452,5 +1452,5 @@
     while (in->point > 0) {
       wchar_t c;
-      char *p = in->buffer + charpos(in,in->point);
+      char *p = in->buffer + charpos(in,in->point-1);
       size_t res = mbrtowc(&c, p, strlen(p), &mbs);
       if (*p && (res <= 0 || !iswalnum (c)))
