--- src/main.c
+++ src/main.c
@@ -868,7 +868,7 @@
 	int prompt_len;
 
 	prompt = strip_ctrl_codes (subshell_prompt);
-	prompt_len = strlen (prompt);
+	prompt_len = mbstrlen (prompt);
 
 	/* Check for prompts too big */
 	if (COLS > 8 && prompt_len > COLS - 8) {
@@ -1844,7 +1844,7 @@
     if (xterm_flag && xterm_title) {
 	p = s = g_strdup (strip_home_and_password (cpanel->cwd));
 	do {
-	    if (!is_printable (*s))
+	    if (*s < ' ')
 		*s = '?';
 	} while (*++s);
 	fprintf (stdout, "\33]0;mc - %s\7", p);
--- src/menu.c
+++ src/menu.c
@@ -20,6 +20,7 @@
 #include <stdarg.h>
 #include <sys/types.h>
 #include <ctype.h>
+#include <wchar.h>
 #include "global.h"
 #include "tty.h"
 #include "menu.h"
--- src/util.c
+++ src/util.c
@@ -84,11 +84,13 @@
     if (SLsmg_Is_Unicode) {
 	static mbstate_t s;
 	int len;
+	const char *str0 = str;
 
 	len = mbsrtowcs (NULL, &str, -1, &s);
 	if (len < 0) {
-	    memset (&s, 0, sizeof (s));
-	    return -1;
+	    	memset (&s, 0, sizeof (s));
+		/* invalid multibyte character, probably not UTF-8 string */
+		return strlen (str0);
 	}
 	return len;
     } else
@@ -917,7 +919,7 @@
 	    continue;
 	}
 
-	if (is_printable(*r))
+	if ((unsigned char)*r >= ' ')
 	    *w++ = *r;
 	++r;
     }
