diff -Nur mc-4.6.0.orig/src/panel.h mc-4.6.0.chng/src/panel.h
--- mc-4.6.0.orig/src/panel.h	Thu Nov 14 08:25:19 2002
+++ mc-4.6.0.chng/src/panel.h	Thu Feb  6 20:31:24 2003
@@ -6,7 +6,7 @@
 
 #define selection(p) (&(p->dir.list[p->selected]))
 #define other_panel get_other_panel()
-#define DEFAULT_USER_FORMAT "half type name | size | perm"
+#define DEFAULT_USER_FORMAT "half type,name,|,mtime,|,perm"
 
 #define LIST_TYPES	4
 
