diff -urN mc-4.5.52.org/src/dir.c mc-4.5.52/src/dir.c
--- mc-4.5.52.org/src/dir.c	Mon Mar 19 20:44:47 2001
+++ mc-4.5.52/src/dir.c	Mon Mar 19 20:55:42 2001
@@ -44,7 +44,7 @@
 #endif
 
 /* If true show files ending in ~ */
-int show_backups = 0;
+int show_backups = 1;
 
 /* If false then directories are shown separately from files */
 int mix_all_files = 0;
diff -urN mc-4.5.52.org/src/panel.h mc-4.5.52/src/panel.h
--- mc-4.5.52.org/src/panel.h	Mon Mar 19 20:44:47 2001
+++ mc-4.5.52/src/panel.h	Mon Mar 19 20:55:46 2001
@@ -292,7 +292,7 @@
 
 void try_to_select (WPanel *panel, char *name);
 
-#define DEFAULT_USER_FORMAT "half type,name,|,size,|,perm"
+#define DEFAULT_USER_FORMAT "half type,name,|,mtime,|,perm"
 
 /* This were in main: */
 void unmark_files (WPanel *panel);
