diff -urN mc-4.5.52.org/lib/mc.ext.in mc-4.5.52/lib/mc.ext.in
--- mc-4.5.52.org/lib/mc.ext.in	Mon Mar 19 20:44:46 2001
+++ mc-4.5.52/lib/mc.ext.in	Mon Mar 19 20:53:54 2001
@@ -350,7 +350,7 @@
 regex/\.([Dd]oc|DOC)$
 	#View=%view{ascii} catdoc -w %f
 	View=%view{ascii} word2x -f text %f -
-type/Microsoft Word Document
+type/Microsoft\ Word\ Document
 	#View=%view{ascii} catdoc -w %f
 	View=%view{ascii} word2x -f text %f -
  
@@ -377,7 +377,7 @@
 	Start Emacs on this RMAIL file=emacs %f
 	Open=emacs %f
 
-type/mail
+type/Mail
 	Open=elm -f %f
 	View=%view{ascii} mcmfmt < %f
 
@@ -496,14 +496,14 @@
 	Extract=cat '%f' | cpio -ic
 
 # gzip
-type/gzip
+type/gzip\ compressed\ data
 	Open=gzip -dc %f | %var{PAGER:more}
 	View=%view{ascii} gzip -dc %f 2>/dev/null
 	Edit=I=`date +%%s`; export I; gzip -cd %f >/tmp/gzed.$I && %var{EDITOR:vi} /tmp/gzed.$I && gzip -c /tmp/gzed.$I > %f; rm -f /tmp/gzed.$I
 	Uncompress=gunzip %f
 
 # bzip2
-type/bzip2
+type/bzip2\ compressed\ data
 	Open=bzip2 -dc %f | %var{PAGER:more}
 	View=%view{ascii} bzip2 -dc %f 2>/dev/null
 	Edit=I=`date +%%s`; export I; bzip2 -cd %f >/tmp/bzed.$I && %var{EDITOR:vi} /tmp/bzed.$I && bzip2 -c /tmp/bzed.$I > %f; rm -f /tmp/bzed.$I
@@ -517,7 +517,7 @@
 	Uncompress=bunzip %f
 
 # compress
-type/compress
+type/compress\'d\ data
 	Open=gzip -dc %f | %var{PAGER:more}
 	View=%view{ascii} gzip -dc %f 2>/dev/null
 	Edit=I=`date +%%s`; export I; gzip -cd %f >/tmp/gzed.$I && %var{EDITOR:vi} /tmp/gzed.$I && gzip -c /tmp/gzed.$I > %f; rm -f /tmp/gzed.$I
