diff -ruN mc-4.5.51.orig/src/user.c mc-4.5.51/src/user.c
--- mc-4.5.51.orig/src/user.c	Mon Jul  3 16:32:08 2000
+++ mc-4.5.51/src/user.c	Wed Oct  4 22:21:56 2000
@@ -208,8 +208,6 @@
     case 'b': return strip_ext((*quote_func) (fname, 0));
     case 'x': return (*quote_func) (extension(fname), 0);
     case 'd': return (*quote_func) (panel->cwd, 0);
-    case 's': if (!panel->marked)
-		return (*quote_func) (fname, 0);
     case 'i': /* indent equal number cursor position in line */
         if (s_editwidget)
 	        return g_strnfill (s_editwidget->curs_col, ' ');
@@ -229,6 +227,8 @@
 	if (menu)
 	    return (*quote_func) (menu, 0);
 	break;
+    case 's': if (!panel->marked)
+		return (*quote_func) (fname, 0);
 
 	/* Fall through */
 
