--- mc-4.5.51/gnome/gaction.c.gnome	Mon Jul  3 10:32:12 2000
+++ mc-4.5.51/gnome/gaction.c	Thu Aug 17 17:48:28 2000
@@ -234,10 +234,13 @@
 		g_free (quoted);
 		g_free (editor_cmd);
 	} else {
-		char *cmd = g_strconcat (editor, " %s", NULL);
+    char *quoted = name_quote (fname, 0);
+		char *cmd = g_strconcat (editor, " ", quoted , NULL);
 
 		gmc_execute (fname, cmd, 0);
 		g_free (cmd);
+    g_free (quoted);
+
 	}
 
 	g_free (editor);
