diff -Nur mc-4.5.55.orig/gtkedit/edit-widget.h mc-4.5.55.chng/gtkedit/edit-widget.h
--- mc-4.5.55.orig/gtkedit/edit-widget.h	Tue Jun 26 17:52:12 2001
+++ mc-4.5.55.chng/gtkedit/edit-widget.h	Fri May 31 14:02:59 2002
@@ -133,12 +133,11 @@
 
 typedef struct editor_widget WEdit;
 
-#define EDIT_DIR           "/.cedit"
-#define SYNTAX_FILE        "/.cedit/Syntax"
-#define CLIP_FILE          "/.cedit/cooledit.clip"
-#define MACRO_FILE         "/.cedit/cooledit.macros"
-#define BLOCK_FILE         "/.cedit/cooledit.block"
-#define ERROR_FILE         "/.cedit/cooledit.error"
-#define TEMP_FILE          "/.cedit/cooledit.temp"
+#define SYNTAX_FILE        "Syntax"
+#define CLIP_FILE          "cooledit.clip"
+#define MACRO_FILE         "cooledit.macros"
+#define BLOCK_FILE         "cooledit.block"
+#define ERROR_FILE         "cooledit.error"
+#define TEMP_FILE          "cooledit.temp"
 
 #endif
diff -Nur mc-4.5.55.orig/gtkedit/edit.c mc-4.5.55.chng/gtkedit/edit.c
--- mc-4.5.55.orig/gtkedit/edit.c	Sun Aug 12 20:02:56 2001
+++ mc-4.5.55.chng/gtkedit/edit.c	Fri May 31 14:07:37 2002
@@ -2768,14 +2768,14 @@
     int nomark;
     struct stat status;
     long start_mark, end_mark;
-    char *block_file = catstrs (home_dir, BLOCK_FILE, 0);
-    char *error_file = catstrs (home_dir, ERROR_FILE, 0);
+    char *block_file = catstrs (home_dir, home_cedit, BLOCK_FILE, 0);
+    char *error_file = catstrs (home_dir, home_cedit, ERROR_FILE, 0);
     int rc = 0;
 
     nomark = eval_marks (edit, &start_mark, &end_mark); 
     if (! nomark) /* remember marked or not */
-	edit_save_block (edit, block_file = catstrs (home_dir, BLOCK_FILE, 0),
-							start_mark, end_mark);
+	edit_save_block (edit, block_file = 
+	    catstrs (home_dir, home_cedit, BLOCK_FILE, 0), start_mark, end_mark);
 
     /* run shell scripts from menu */     
     user_menu_cmd (edit);
diff -Nur mc-4.5.55.orig/gtkedit/edit.h mc-4.5.55.chng/gtkedit/edit.h
--- mc-4.5.55.orig/gtkedit/edit.h	Sun Aug 19 19:01:01 2001
+++ mc-4.5.55.chng/gtkedit/edit.h	Fri May 31 14:11:39 2002
@@ -661,6 +661,7 @@
 #endif				/* ! MIDNIGHT */
 
 extern char *home_dir;
+extern char *home_cedit;
 
 #define NUM_SELECTION_HISTORY 64
 
diff -Nur mc-4.5.55.orig/gtkedit/editcmd.c mc-4.5.55.chng/gtkedit/editcmd.c
--- mc-4.5.55.orig/gtkedit/editcmd.c	Thu Aug 23 19:21:13 2001
+++ mc-4.5.55.chng/gtkedit/editcmd.c	Fri May 31 14:09:13 2002
@@ -680,7 +680,7 @@
 {
     char *filename;
     int file;
-    filename = catstrs (home_dir, MACRO_FILE, 0);
+    filename = catstrs (home_dir, home_cedit, MACRO_FILE, 0);
     if ((file = open (filename, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) == -1)
 	return 0;
     close (file);
@@ -715,7 +715,7 @@
     if (saved_macros_loaded)
 	if ((j = macro_exists (k)) < 0)
 	    return 0;
-    g = fopen (catstrs (home_dir, TEMP_FILE, 0), "w");
+    g = fopen (catstrs (home_dir, home_cedit, TEMP_FILE, 0), "w");
     if (!g) {
 /* This heads the delete macro error dialog box */
 	edit_error_dialog (_(" Delete macro "),
@@ -749,7 +749,8 @@
     }
     fclose (f);
     fclose (g);
-    if (rename (catstrs (home_dir, TEMP_FILE, 0), catstrs (home_dir, MACRO_FILE, 0)) == -1) {
+    if (rename (catstrs (home_dir, home_cedit, TEMP_FILE, 0), 
+	catstrs (home_dir, home_cedit, MACRO_FILE, 0)) == -1) {
 /* This heads the delete macro error dialog box */
 	edit_error_dialog (_(" Delete macro "),
 	   get_sys_error (_(" Error trying to overwrite macro file ")));
@@ -2585,7 +2586,7 @@
 /* copies a block to clipboard file */
 static int edit_save_block_to_clip_file (WEdit * edit, long start, long finish)
 {
-    return edit_save_block (edit, catstrs (home_dir, CLIP_FILE, 0), start, finish);
+    return edit_save_block (edit, catstrs (home_dir, home_cedit, CLIP_FILE, 0), start, finish);
 }
 
 #ifndef MIDNIGHT
@@ -2804,7 +2805,7 @@
 
 void edit_paste_from_X_buf_cmd (WEdit * edit)
 {
-    edit_insert_file (edit, catstrs (home_dir, CLIP_FILE, 0));
+    edit_insert_file (edit, catstrs (home_dir, home_cedit, CLIP_FILE, 0));
 }
 
 #endif /* MIDNIGHT */
@@ -2849,7 +2850,7 @@
     char *exp;
     if (eval_marks (edit, &start_mark, &end_mark))
 	return 1;
-    exp = edit_get_save_file (edit->dir, catstrs (home_dir, CLIP_FILE, 0), _ (" Save Block "));
+    exp = edit_get_save_file (edit->dir, catstrs (home_dir, home_cedit, CLIP_FILE, 0), _ (" Save Block "));
     edit_push_action (edit, KEY_PRESS + edit->start_display);
     if (exp) {
 	if (!*exp) {
@@ -2874,7 +2875,7 @@
 /* returns 1 on success */
 int edit_insert_file_cmd (WEdit * edit)
 {
-    char *exp = edit_get_load_file (edit->dir, catstrs (home_dir, CLIP_FILE, 0), _ (" Insert File "));
+    char *exp = edit_get_load_file (edit->dir, catstrs (home_dir, home_cedit, CLIP_FILE, 0), _ (" Insert File "));
     edit_push_action (edit, KEY_PRESS + edit->start_display);
     if (exp) {
 	if (!*exp) {
@@ -2909,7 +2910,7 @@
 	edit_error_dialog (_(" Sort block "), _(" You must first highlight a block of text. "));
 	return 0;
     }
-    edit_save_block (edit, catstrs (home_dir, BLOCK_FILE, 0), start_mark, end_mark);
+    edit_save_block (edit, catstrs (home_dir, home_cedit, BLOCK_FILE, 0), start_mark, end_mark);
 
     exp = old ? old : "";
 
@@ -2922,7 +2923,9 @@
 	free (old);
     old = exp;
 
-    e = system (catstrs (" sort ", exp, " ", home_dir, BLOCK_FILE, " > ", home_dir, TEMP_FILE, 0));
+    e = system (catstrs (" sort ", exp, " ", 
+	home_dir, home_cedit, BLOCK_FILE, " > ", 
+	home_dir, home_cedit, TEMP_FILE, 0));
     if (e) {
 	if (e == -1 || e == 127) {
 	    edit_error_dialog (_(" Sort "), 
@@ -2940,7 +2943,7 @@
 
     if (edit_block_delete_cmd (edit))
 	return 1;
-    edit_insert_file (edit, catstrs (home_dir, TEMP_FILE, 0));
+    edit_insert_file (edit, catstrs (home_dir, home_cedit, TEMP_FILE, 0));
     return 0;
 }
 
@@ -2957,9 +2960,9 @@
     FILE *block_file = NULL;
 
     char *o = catstrs (mc_home, shell_cmd, 0);   /* original source script */
-    char *h = catstrs (home_dir, EDIT_DIR, shell_cmd, 0); /* home script */
-    char *b = catstrs (home_dir, BLOCK_FILE, 0); /* block file */
-    char *e = catstrs (home_dir, ERROR_FILE, 0); /* error file */
+    char *h = catstrs (home_dir, home_cedit, shell_cmd, 0); /* home script */
+    char *b = catstrs (home_dir, home_cedit, BLOCK_FILE, 0); /* block file */
+    char *e = catstrs (home_dir, home_cedit, ERROR_FILE, 0); /* error file */
     
     if (! (script_home = fopen (h, "r"))) {
 	if (! (script_home = fopen (h, "w"))) {
@@ -2996,12 +2999,12 @@
         edit_save_block (edit, b, start_mark, end_mark);
         
 	/* run your script */
-        execute (catstrs (home_dir, EDIT_DIR, shell_cmd, " ", 
-	    edit->filename, " ", home_dir, BLOCK_FILE, " ", 
-	    home_dir, ERROR_FILE, 0));
+        execute (catstrs (home_dir, home_cedit, shell_cmd, " ", 
+	    edit->filename, " ", home_dir, home_cedit, BLOCK_FILE, " ", 
+	    home_dir, home_cedit, ERROR_FILE, 0));
 
     } else { /* for missing marked block run ... */
-	execute (catstrs (EDIT_DIR, shell_cmd));
+	execute (catstrs ("/", home_cedit, shell_cmd));
     }
 
     edit_refresh_cmd (edit);
diff -Nur mc-4.5.55.orig/gtkedit/editwidget.c mc-4.5.55.chng/gtkedit/editwidget.c
--- mc-4.5.55.orig/gtkedit/editwidget.c	Sat Jun 16 00:06:11 2001
+++ mc-4.5.55.chng/gtkedit/editwidget.c	Fri May 31 14:00:28 2002
@@ -313,7 +313,7 @@
 	w->label = (char *) strdup ("");
 
     if (!made_directory) {
-	mkdir (catstrs (home_dir, EDIT_DIR, 0), 0700);
+	mkdir (catstrs (home_dir, home_cedit, 0), 0700);
 	made_directory = 1;
     }
     e = w->editor = CMalloc (sizeof (WEdit));
@@ -1009,7 +1009,7 @@
 	memcpy (option_backup_ext, (char *) &option_backup_ext_int, sizeof (int));
     }
     if (!made_directory) {
-	mkdir (catstrs (home_dir, EDIT_DIR, 0), 0700);
+	mkdir (catstrs (home_dir, home_cedit, 0), 0700);
 	made_directory = 1;
     }
     if (_file) {
diff -Nur mc-4.5.55.orig/gtkedit/gtkedit.c mc-4.5.55.chng/gtkedit/gtkedit.c
--- mc-4.5.55.orig/gtkedit/gtkedit.c	Sun Aug 19 19:26:03 2001
+++ mc-4.5.55.chng/gtkedit/gtkedit.c	Fri May 31 14:37:06 2002
@@ -789,22 +789,44 @@
 
 static void get_home_dir (void)
 {
+    char *buffor;
+
     if (home_dir)		/* already been set */
 	return;
+    
+    buffor = getenv("CONFIG_DIR");
+    if (buffor) {
+	if (buffor[strlen(buffor) - 1] == '/')
+	    home_cedit = strdup (catstrs (buffor, "/cedit/", 0));
+	else
+	    home_cedit = strdup (catstrs (buffor, "cedit/", 0));
+	    	    
+    } else 
+	home_cedit = strdup (".cedit/");
+	
     home_dir = getenv ("HOME");
     if (home_dir)
-	if (*home_dir) {
-	    home_dir = strdup (home_dir);
-	    return;
-	}
-    home_dir = (getpwuid (geteuid ()))->pw_dir;
-    if (home_dir)
-	if (*home_dir) {
+	if (*home_dir) 
 	    home_dir = strdup (home_dir);
-	    return;
-	}
-    fprintf (stderr, _("gtkedit.c: HOME environment variable not set and no passwd entry - aborting\n"));
-    abort ();
+
+    if (!home_dir) {
+        home_dir = (getpwuid (geteuid ()))->pw_dir;
+	if (home_dir)
+	    if (*home_dir)
+		home_dir = strdup (home_dir);
+    }
+    
+    if (!home_dir) {
+        fprintf (stderr, _("gtkedit.c: HOME environment variable not set and no passwd entry - aborting\n"));
+	abort ();
+    }
+
+    if (home_dir[strlen(home_dir) - 1] != '/') {
+	buffor = home_dir;
+	home_dir = strdup (catstrs (home_dir, "/", 0));
+	free (buffor);
+    }    
+    
 }
 
 static gchar *gtk_edit_get_chars (GtkEditable * editable,
@@ -930,7 +952,7 @@
     gtk_edit_configure_font_dimensions (edit);
 
     if (!made_directory) {
-	mkdir (catstrs (home_dir, EDIT_DIR, 0), 0700);
+	mkdir (catstrs (home_dir, home_cedit, 0), 0700);
 	made_directory = 1;
     }
     GTK_EDITABLE (edit)->editable = TRUE;
diff -Nur mc-4.5.55.orig/gtkedit/syntax.c mc-4.5.55.chng/gtkedit/syntax.c
--- mc-4.5.55.orig/gtkedit/syntax.c	Fri May 31 14:22:10 2002
+++ mc-4.5.55.chng/gtkedit/syntax.c	Fri May 31 14:02:16 2002
@@ -608,7 +608,7 @@
     if (*filename == '/')
 	return fopen (filename, "r");
     strcpy (p, home_dir);
-    strcat (p, EDIT_DIR "/");
+    strcat (p, home_cedit);
     strcat (p, filename);
     syntax_free (error_file_name);
     error_file_name = (char *) strdup (p);
@@ -1533,7 +1533,7 @@
 	if (!*edit->filename && !type)
 	    return;
     }
-    f = catstrs (home_dir, SYNTAX_FILE, 0);
+    f = catstrs (home_dir, home_cedit, SYNTAX_FILE, 0);
     r = edit_read_syntax_file (edit, names, f, edit ? edit->filename : 0, get_first_editor_line (edit), type);
     if (r == -1) {
 	edit_free_syntax_rules (edit);
diff -Nur mc-4.5.55.orig/lib/mc.sh.in mc-4.5.55.chng/lib/mc.sh.in
--- mc-4.5.55.orig/lib/mc.sh.in	Fri May 31 14:22:12 2002
+++ mc-4.5.55.chng/lib/mc.sh.in	Fri May 31 09:04:24 2002
@@ -1,8 +1,14 @@
 mc ()
 {
-        mkdir -p $HOME/.mc/tmp 2> /dev/null
-	chmod 700 $HOME/.mc/tmp
-	MC=$HOME/.mc/tmp/mc-$$
+	if [ "$CONFIG_DIR" == "" ]; then
+	    HOMEMC="$HOME/.mc"
+	else
+	    HOMEMC="$HOME/$CONFIG_DIR/mc"
+	fi
+
+        mkdir -p $HOMEMC/tmp 2> /dev/null
+	chmod 700 $HOMEMC/tmp
+	MC=$HOMEMC/tmp/mc-$$
 	@prefix@/bin/mc -P "$@" > "$MC"
 	cd "`cat $MC`"
 	rm -f "$MC"
diff -Nur mc-4.5.55.orig/src/cmd.c mc-4.5.55.chng/src/cmd.c
--- mc-4.5.55.orig/src/cmd.c	Sun Aug 12 10:36:26 2001
+++ mc-4.5.55.chng/src/cmd.c	Fri May 31 10:18:48 2002
@@ -655,7 +655,7 @@
     extdir = concat_dir_and_file (mc_home, MC_LIB_EXT);
 
     if (dir == 0){
-	buffer = concat_dir_and_file (home_dir, MC_USER_EXT);
+	buffer = g_strconcat (home_dir, home_mc, MC_USER_EXT, NULL);
 	check_for_default (extdir, buffer);
 	do_edit (buffer);
 	g_free (buffer);
@@ -685,12 +685,16 @@
 
     switch (dir) {
 	case 0:
-	    buffer = g_strdup (where ? CEDIT_LOCAL_MENU : MC_LOCAL_MENU);
+	    buffer = g_strconcat (home_dir, config_dir, (where ? CEDIT_LOCAL_MENU : MC_LOCAL_MENU), NULL);
 	    check_for_default (menufile, buffer);
 	    break;
 
 	case 1:
-	    buffer = concat_dir_and_file (home_dir, where ? CEDIT_HOME_MENU : MC_HOME_MENU);
+	    if (where)
+		buffer = g_strconcat(home_dir, home_cedit, CEDIT_HOME_MENU, NULL);
+	    else
+		buffer = g_strconcat(home_dir, home_mc, MC_HOME_MENU, NULL);
+
 	    check_for_default (menufile, buffer);
 	    break;
 	
@@ -1436,7 +1440,7 @@
     
     save_setup ();
     sync_profiles ();
-    str = g_strconcat ( _(" Setup saved to ~/"), PROFILE_NAME, NULL);
+    str = g_strconcat ( _(" Setup saved to ~/"), home_mc, PROFILE_NAME, NULL);
     
     message (0, _(" Setup "), str);
     g_free (str);
diff -Nur mc-4.5.55.orig/src/ext.c mc-4.5.55.chng/src/ext.c
--- mc-4.5.55.orig/src/ext.c	Tue Aug 14 02:55:38 2001
+++ mc-4.5.55.chng/src/ext.c	Fri May 31 09:37:49 2002
@@ -390,7 +390,7 @@
 	int mc_user_ext = 1;
         int home_error = 0;
 
-	extension_file = concat_dir_and_file (home_dir, MC_USER_EXT);
+	extension_file = g_strconcat (home_dir, home_mc, MC_USER_EXT, NULL);
 	if (!exist_file (extension_file)) {
 	    g_free (extension_file);
 check_stock_mc_ext:
@@ -430,8 +430,9 @@
 	if (home_error) {
             char *msg;
             char *msg2;
-            msg = g_strconcat (" ~/", MC_USER_EXT, _(" file error "), NULL);
-            msg2 = g_strconcat (_("Format of the ~/"), MC_USER_EXT, _(" file has changed\n\
+            msg = g_strconcat (" ~/", home_mc, MC_USER_EXT, _(" file error "), NULL);
+            msg2 = g_strconcat (_("Format of the ~/"), home_mc, MC_USER_EXT,
+_(" file has changed\n\
 with version 3.0. You may want either to\n\
 copy it from "), mc_home, _("mc.ext or use that\n\
 file as an example of how to write it.\n\
diff -Nur mc-4.5.55.orig/src/ext.h mc-4.5.55.chng/src/ext.h
--- mc-4.5.55.orig/src/ext.h	Mon Feb 26 22:21:18 2001
+++ mc-4.5.55.chng/src/ext.h	Fri May 31 08:58:29 2002
@@ -14,7 +14,7 @@
 #    define MC_LIB_EXT  "mc.ext"
 #else
 #    ifndef HAVE_GNOME
-#        define MC_USER_EXT ".mc/bindings" 
+#        define MC_USER_EXT "bindings" 
 #        define MC_LIB_EXT  "mc.ext"
 #    endif
 #endif
diff -Nur mc-4.5.55.orig/src/global.h mc-4.5.55.chng/src/global.h
--- mc-4.5.55.orig/src/global.h	Sun Aug 19 18:23:24 2001
+++ mc-4.5.55.chng/src/global.h	Fri May 31 14:30:10 2002
@@ -27,6 +27,9 @@
 #endif
 
 extern char *home_dir;
+extern char *home_mc;
+extern char *home_cedit;
+extern char *config_dir;
 
 #ifdef min
 #undef min
diff -Nur mc-4.5.55.orig/src/hotlist.c mc-4.5.55.chng/src/hotlist.c
--- mc-4.5.55.orig/src/hotlist.c	Mon Jul  9 08:43:45 2001
+++ mc-4.5.55.chng/src/hotlist.c	Fri May 31 09:46:47 2002
@@ -1349,7 +1349,7 @@
     }
 
     if (!hotlist_file_name)
-	hotlist_file_name = concat_dir_and_file (home_dir, HOTLIST_FILENAME);
+	hotlist_file_name = g_strconcat (home_dir, home_mc, HOTLIST_FILENAME);
     
     hotlist	       = new_hotlist ();
     hotlist->type      = HL_TYPE_GROUP;
@@ -1370,9 +1370,9 @@
 	char    *msg;
 
 	msg = g_strconcat (_("Hotlist is now kept in file ~/"),
-			    HOTLIST_FILENAME, "\n",
+			    home_mc, HOTLIST_FILENAME, "\n",
 			    _("MC will load hotlist from ~/"),
-			    PROFILE_NAME, "\n",
+			    home_mc, PROFILE_NAME, "\n",
 			    _("and then delete [Hotlist] section there"),
 			    NULL);
 	message (0, _(" Hotlist Load "), msg);
@@ -1391,7 +1391,7 @@
 	} else {
 	    char *msg;
 
-	    msg = g_strconcat (_("MC was unable to write ~/"), HOTLIST_FILENAME,
+	    msg = g_strconcat (_("MC was unable to write ~/"), home_mc, HOTLIST_FILENAME,
 				_(" file, your old hotlist entries were not deleted"), NULL);
 
 	    message (D_ERROR, _(" Hotlist Load "), msg);
@@ -1406,16 +1406,18 @@
 	    char        *msg;
 
 	    msg = g_strconcat (
-		    _("You have ~/"), HOTLIST_FILENAME, _(" file and [Hotlist] section in ~/"), PROFILE_NAME, "\n",
-		    _("Your ~/"), HOTLIST_FILENAME, _(" most probably was created\n"),
+		    _("You have ~/"), home_mc, HOTLIST_FILENAME, _(" file and [Hotlist] section in ~/"), 
+		    home_mc, PROFILE_NAME, "\n",
+		    _("Your ~/"), home_mc, HOTLIST_FILENAME, _(" most probably was created\n"),
 		    _("by an earlier development version of MC\nand is more actual than ~/"),
-		    PROFILE_NAME, _(" entries\n\n"),
+		    home_mc, PROFILE_NAME, _(" entries\n\n"),
 		    _("You can choose between\n\n"
-		      "  Remove - remove old hotlist entries from ~/"), PROFILE_NAME, "\n",
+		      "  Remove - remove old hotlist entries from ~/"), 
+		    home_mc, PROFILE_NAME, "\n",
 		    _("  Keep   - keep your old entries; you will be asked\n"
 		      "           the same question next time\n"
 		      "  Merge  - add old entries to hotlist as group \"Entries from ~/"),
-		    PROFILE_NAME, "\"\n\n", NULL);
+		    home_mc, PROFILE_NAME, "\"\n\n", NULL);
 
 	    result = query_dialog (_(" Hotlist Load "),
 				   msg, D_ERROR, 3, _("&Remove"), _("&Keep"), _("&Merge"));
@@ -1430,7 +1432,7 @@
 
 		old            = new_hotlist ();
 		old->type      = HL_TYPE_GROUP;
-		old->label     = g_strconcat (_(" Entries from ~/"), PROFILE_NAME, NULL);
+		old->label     = g_strconcat (_(" Entries from ~/"), home_mc, PROFILE_NAME, NULL);
 		old->up	       = hotlist;
 		old->head      = hotlist->head;
 		old->next      = grp;
@@ -1439,7 +1441,7 @@
 		if (!save_hotlist ()){
 		    char *str;
 
-		    str = g_strconcat (_("MC was unable to write ~/"), HOTLIST_FILENAME,
+		    str = g_strconcat (_("MC was unable to write ~/"), home_mc, HOTLIST_FILENAME,
 					_(" file your old hotlist entries were not deleted"), NULL);
 
 		    message (D_ERROR, _(" Hotlist Load "), str);
diff -Nur mc-4.5.55.orig/src/main.c mc-4.5.55.chng/src/main.c
--- mc-4.5.55.orig/src/main.c	Fri May 31 14:22:11 2002
+++ mc-4.5.55.chng/src/main.c	Fri May 31 17:34:26 2002
@@ -307,6 +307,9 @@
 
 /* The home directory */
 char *home_dir;
+char *home_mc;
+char *home_cedit;
+char *config_dir;
 
 #ifndef HAVE_X
 /* The value of the other directory, only used when loading the setup */
@@ -2922,15 +2925,20 @@
  * Previous versions of the program had all of their files in
  * the $HOME, we are now putting them in $HOME/.mc
  */
+ 
 #ifdef OS2_NT
 #    define compatibility_move_mc_files()
 #else
 
+/*
+ *   This function should be fixed to provide new CONFIG_DIR
+ *   functionality (pascalek@pld.org.pl)
+ */
 static int
 do_mc_filename_rename (char *mc_dir, char *o_name, char *n_name)
 {
 	char *full_o_name = concat_dir_and_file (home_dir, o_name);
-	char *full_n_name = g_strconcat (home_dir, MC_BASE, n_name, NULL);
+	char *full_n_name = g_strconcat (home_dir, home_mc, n_name, NULL);
 	int move;
 	
 	move = 0 == rename (full_o_name, full_n_name);
@@ -2971,7 +2979,7 @@
 static void
 compatibility_move_mc_files (void)
 {
-	char *mc_dir = concat_dir_and_file (home_dir, ".mc");
+	char *mc_dir = g_strconcat (home_dir, home_mc);
 	
 	do_compatibility_move (mc_dir);
 	g_free (mc_dir);
@@ -2981,6 +2989,8 @@
 int
 main (int argc, char *argv [])
 {
+    char *buffer;
+    
     /* We had LC_CTYPE before, LC_ALL includs LC_TYPE as well */
     setlocale (LC_ALL, "");
     bindtextdomain ("mc", LOCALEDIR);
@@ -2994,7 +3004,6 @@
 	    printf ("GMC IS WAITING %d\n", getpid ());
 	    while (!i)
 		    ;
-
     }
 #endif
     /* Initialize list of all user group for timur_clr_mode */
@@ -3002,15 +3011,35 @@
     
     OS_Setup ();
 
-    /* This variable is used by the subshell */
-    home_dir = getenv ("HOME");
-    if (!home_dir) {
+    buffer = getenv("CONFIG_DIR");
+    if (buffer) {
+        if (buffer[strlen(buffer) - 1] == PATH_SEP)
+	    config_dir = buffer;
+        else
+    	    config_dir = g_strconcat (buffer, PATH_SEP_STR, NULL);
+
+	home_mc = g_strconcat (config_dir, "mc/", NULL);
+	home_cedit = g_strconcat (config_dir, "cedit/", NULL);
+    } else {
+	home_mc = g_strdup (".mc/");
+	home_cedit = g_strdup (".cedit/");
+	config_dir = g_strdup ("");
+    }
+
+    /* This variable is used by the subshell */    
+    buffer = getenv("HOME");
+    if (!buffer) {
 	    /* mc_home was computed by OS_Setup */
-	    home_dir = mc_home; 
+	    buffer = mc_home; 
     }
-    
+
+    if (buffer[strlen(buffer) - 1] == PATH_SEP)
+	home_dir = g_strdup (buffer);
+    else
+	home_dir = g_strconcat (buffer, PATH_SEP_STR, NULL);
+
     vfs_init ();
-    
+
 #ifdef HAVE_X
     /* NOTE: This call has to be before any our argument handling :) */
 
@@ -3209,6 +3238,9 @@
 #ifdef _OS_NT
     /* On NT, home_dir is malloced */
     g_free (home_dir);
+    g_free (home_mc);
+    g_free (home_cedit);
+    g_free (config_dir);
 #endif
 #if defined(OS2_NT)
     if (print_last_wd == 2){
diff -Nur mc-4.5.55.orig/src/main.h mc-4.5.55.chng/src/main.h
--- mc-4.5.55.orig/src/main.h	Tue Aug 21 08:32:41 2001
+++ mc-4.5.55.chng/src/main.h	Fri May 31 07:38:16 2002
@@ -193,12 +193,6 @@
 void init_menu (void);
 void exec_shell (void);
 
-#ifdef OS2_NT
-#    define MC_BASE ""
-#else
-#    define MC_BASE "/.mc/"
-#endif /* !OS2_NT */
-
 /* Back hack to define the following routines only if the client code
  * has included panel.h
  */
diff -Nur mc-4.5.55.orig/src/poptconfig.c mc-4.5.55.chng/src/poptconfig.c
--- mc-4.5.55.orig/src/poptconfig.c	Tue Aug  3 07:17:47 1999
+++ mc-4.5.55.chng/src/poptconfig.c	Fri May 31 13:00:21 2002
@@ -125,6 +125,10 @@
     return 0;
 }
 
+/*
+ *  It should be rewriten to meet home_etc requirements
+ *  after deep look insinde popt code... (pascalek@pld.org.pl)
+ */
 int poptReadDefaultConfig(poptContext con, int useEnv) {
     char * fn, * home;
     int rc;
diff -Nur mc-4.5.55.orig/src/setup.c mc-4.5.55.chng/src/setup.c
--- mc-4.5.55.orig/src/setup.c	Thu Aug 16 23:01:12 2001
+++ mc-4.5.55.chng/src/setup.c	Fri May 31 09:43:54 2002
@@ -348,7 +348,7 @@
     int  i;
     char buffer [BUF_TINY];
 
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = g_strconcat (home_dir, home_mc, PROFILE_NAME, NULL);
 
     /* Save integer options */
     for (i = 0; layout [i].opt_name; i++){
@@ -366,7 +366,7 @@
     char *profile;
     int  i;
 
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = g_strconcat (home_dir, home_mc, PROFILE_NAME, NULL);
 
     /* Save integer options */
     for (i = 0; options [i].opt_name; i++)
@@ -415,7 +415,7 @@
     char *profile;
 
     saving_setup = 1;
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = g_strconcat (home_dir, home_mc, PROFILE_NAME, NULL);
 
     save_configure ();
 
@@ -560,7 +560,7 @@
     if (profile_name)
 	    return profile_name;
 
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = g_strconcat (home_dir, home_mc, PROFILE_NAME, NULL);
     if (!exist_file (profile)){
 	inifile = concat_dir_and_file (mc_home, "mc.ini");
 	if (exist_file (inifile)){
diff -Nur mc-4.5.55.orig/src/setup.h mc-4.5.55.chng/src/setup.h
--- mc-4.5.55.orig/src/setup.h	Tue Mar 30 08:10:26 1999
+++ mc-4.5.55.chng/src/setup.h	Fri May 31 08:59:09 2002
@@ -32,7 +32,7 @@
 #    define PROFILE_NAME     "mc.ini"
 #    define HOTLIST_FILENAME "mc.hot"
 #else
-#    define PROFILE_NAME     ".mc/ini"
-#    define HOTLIST_FILENAME ".mc/hotlist"
+#    define PROFILE_NAME     "ini"
+#    define HOTLIST_FILENAME "hotlist"
 #endif
 #endif
diff -Nur mc-4.5.55.orig/src/subshell.c mc-4.5.55.chng/src/subshell.c
--- mc-4.5.55.orig/src/subshell.c	Fri May 31 14:22:10 2002
+++ mc-4.5.55.chng/src/subshell.c	Fri May 31 10:51:08 2002
@@ -214,6 +214,8 @@
 {
     int pty_slave;
     char *init_file = NULL;
+    char *buffor;
+    char *inputrc;
 	
     setsid ();  /* Get a fresh terminal session */
 
@@ -282,23 +284,33 @@
     switch (subshell_type)
     {
         case BASH:
-	    init_file = ".mc/bashrc";
+	    init_file = g_strconcat (home_mc, "bashrc", NULL);
 	    if (access (init_file, R_OK) == -1)
-		init_file = ".bashrc";
+		init_file = g_strdup(".bashrc");
 
 	    /* Make MC's special commands not show up in bash's history */
 	    putenv ("HISTCONTROL=ignorespace");
 
 	    /* Allow alternative readline settings for MC */
-	    if (access (".mc/inputrc", R_OK) == 0)
-		putenv ("INPUTRC=.mc/inputrc");
-
+	    inputrc = g_strconcat (home_mc, "inputrc", NULL);
+	    if (access (inputrc, R_OK) == 0) {
+		buffor = g_strconcat("INPUTRC=", home_mc, NULL);
+		putenv (buffor);
+		g_free(buffor);
+	    }
+	    g_free(inputrc);
+	
 	    break;
 
 	case TCSH:
-	    init_file = ".mc/tcshrc";
-	    if (access (init_file, R_OK) == -1)
-		init_file += 3;
+	    /*
+	     *  UNUSED VARIABLE?! (pascalek@pld.org.pl)
+	     */
+	    init_file = g_strconcat (home_mc, "tcshrc", NULL);
+	    if (access (init_file, R_OK) == -1) {
+		g_free (init_file);
+		init_file = g_strdup ("/tcshrc");
+	    }
 	break;
 
 	case ZSH:
@@ -331,6 +343,7 @@
     {
         case BASH:
 	    execl (shell, "bash", "-rcfile", init_file, NULL);
+	    g_free (init_file);
 	    break;
 
         case TCSH:
diff -Nur mc-4.5.55.orig/src/tree.c mc-4.5.55.chng/src/tree.c
--- mc-4.5.55.orig/src/tree.c	Tue Aug 14 02:55:38 2001
+++ mc-4.5.55.chng/src/tree.c	Fri May 31 10:31:40 2002
@@ -191,7 +191,7 @@
     error = tree_store_save ();
 
     if (error){
-	fprintf (stderr, _("Cannot open the %s file for writing:\n%s\n"), MC_TREE,
+	fprintf (stderr, _("Cannot open the %s%s file for writing:\n%s\n"), home_mc, MC_TREE,
 		 unix_error_string (error));
 	return;
     }
diff -Nur mc-4.5.55.orig/src/treestore.c mc-4.5.55.chng/src/treestore.c
--- mc-4.5.55.orig/src/treestore.c	Tue Aug 14 02:55:38 2001
+++ mc-4.5.55.chng/src/treestore.c	Fri May 31 10:32:57 2002
@@ -283,7 +283,7 @@
 	char *name;
 	int retval;
 
-	name = concat_dir_and_file (home_dir, MC_TREE);
+	name = g_strconcat (home_dir, home_mc, MC_TREE, NULL);
 	retval = tree_store_load_from (name);
 	g_free (name);
 
@@ -386,7 +386,7 @@
 	char *name;
 	int retval;
 
-	tmp = concat_dir_and_file (home_dir, MC_TREE_TMP);
+	tmp = g_strconcat (home_dir, home_mc, MC_TREE_TMP, NULL);
 	retval = tree_store_save_to (tmp);
 
 	if (retval) {
@@ -394,7 +394,7 @@
 		return retval;
 	}
 
-	name = concat_dir_and_file (home_dir, MC_TREE);
+	name = g_strconcat (home_dir, home_mc, MC_TREE, NULL);
 	retval = rename (tmp, name);
 
 	g_free (tmp);
diff -Nur mc-4.5.55.orig/src/treestore.h mc-4.5.55.chng/src/treestore.h
--- mc-4.5.55.orig/src/treestore.h	Tue Sep 14 02:31:35 1999
+++ mc-4.5.55.chng/src/treestore.h	Fri May 31 08:59:59 2002
@@ -7,8 +7,8 @@
 #   define MC_TREE "mcn.tre"
 #   define MC_TREE_TMP "mcn.tr~"
 #else
-#   define MC_TREE ".mc/Tree"
-#   define MC_TREE_TMP ".mc/Tree.tmp"
+#   define MC_TREE "Tree"
+#   define MC_TREE_TMP "Tree.tmp"
 #endif
 
 typedef struct tree_entry {
diff -Nur mc-4.5.55.orig/src/user.c mc-4.5.55.chng/src/user.c
--- mc-4.5.55.orig/src/user.c	Tue Aug 14 02:55:38 2001
+++ mc-4.5.55.chng/src/user.c	Fri May 31 10:20:25 2002
@@ -673,11 +673,15 @@
 	return;
     }
     
-    menu = g_strdup (edit_widget ? CEDIT_LOCAL_MENU : MC_LOCAL_MENU);
+    menu = g_strconcat (config_dir, (edit_widget ? CEDIT_LOCAL_MENU : MC_LOCAL_MENU), NULL);
     if (!exist_file (menu) || !menu_file_own (menu)){
 	g_free (menu);
-        menu = concat_dir_and_file \
-                            (home_dir, edit_widget ? CEDIT_HOME_MENU : MC_HOME_MENU);
+	
+	if (edit_widget)
+	    menu = g_strconcat(home_dir, home_cedit, CEDIT_HOME_MENU, NULL);
+	else
+	    menu = g_strconcat(home_dir, home_mc, MC_HOME_MENU, NULL);
+
 	if (!exist_file (menu)){
 	    g_free (menu);
 	    menu = concat_dir_and_file \
diff -Nur mc-4.5.55.orig/src/user.h mc-4.5.55.chng/src/user.h
--- mc-4.5.55.orig/src/user.h	Tue Jul 24 00:50:55 2001
+++ mc-4.5.55.chng/src/user.h	Fri May 31 10:15:03 2002
@@ -25,11 +25,11 @@
 #   define MC_HINT              "mc.hnt"
 #else
 #   define CEDIT_GLOBAL_MENU    "cedit.menu"
-#   define CEDIT_LOCAL_MENU     ".cedit.menu"
-#   define CEDIT_HOME_MENU      ".cedit/menu"
+#   define CEDIT_LOCAL_MENU     "cedit.menu"
+#   define CEDIT_HOME_MENU      "menu"
 #   define MC_GLOBAL_MENU       "mc.menu"
-#   define MC_LOCAL_MENU        ".mc.menu"
-#   define MC_HOME_MENU         ".mc/menu"
+#   define MC_LOCAL_MENU        "mc.menu"
+#   define MC_HOME_MENU         "menu"
 #   define MC_HINT              "mc.hint"
 #endif
 
diff -Nur mc-4.5.55.orig/src/widget.c mc-4.5.55.chng/src/widget.c
--- mc-4.5.55.orig/src/widget.c	Tue Jul 31 13:04:21 2001
+++ mc-4.5.55.chng/src/widget.c	Fri May 31 10:35:00 2002
@@ -896,7 +896,7 @@
 	return 0;
     if (!*input_name)
 	return 0;
-    profile = concat_dir_and_file (home_dir, HISTORY_FILE_NAME);
+    profile = g_strconcat (home_dir, home_mc, HISTORY_FILE_NAME, NULL);
     for (i = 0;; i++) {
 	char key_name[BUF_TINY];
 	char this_entry[BUF_LARGE];
@@ -933,7 +933,7 @@
     if (!num_history_items_recorded)	/* this is how to disable */
 	return;
 
-    profile = concat_dir_and_file (home_dir, HISTORY_FILE_NAME);
+    profile = g_strconcat (home_dir, home_mc, HISTORY_FILE_NAME, NULL);
 
     if ((i = open (profile, O_CREAT | O_EXCL, S_IRUSR | S_IWUSR)) != -1)
 	close (i);
diff -Nur mc-4.5.55.orig/src/widget.h mc-4.5.55.chng/src/widget.h
--- mc-4.5.55.orig/src/widget.h	Tue Jul 31 13:04:21 2001
+++ mc-4.5.55.chng/src/widget.h	Fri May 31 09:01:59 2002
@@ -90,7 +90,7 @@
 
 /* For history load-save functions */
 #define INPUT_LAST_TEXT ((char *) 2)
-#define HISTORY_FILE_NAME ".mc/history"
+#define HISTORY_FILE_NAME "history"
 
 typedef struct {
     Widget widget;
diff -Nur mc-4.5.55.orig/vfs/extfs/uzoo.in mc-4.5.55.chng/vfs/extfs/uzoo.in
--- mc-4.5.55.orig/vfs/extfs/uzoo.in	Thu Sep 21 22:58:11 2000
+++ mc-4.5.55.chng/vfs/extfs/uzoo.in	Fri May 31 06:14:27 2002
@@ -9,6 +9,12 @@
 # Now it is 'only' very ugly (it creates temporary files in ~/.mc/
 #
 
+if [ "$CONFIG_DIR" != "" ]; then
+    HOMEMC="$HOME/$CONFIG_DIR/mc"
+else
+    HOMEMC="$HOME/.mc"
+fi
+
 ZOO=zoo
 
 mczoofs_list ()
@@ -46,9 +52,9 @@
 if echo $2 | grep '\.zoo$' >/dev/null; then
   :
 else
-  SYMLINK=$HOME/.mc/temporary.$2.zoo
+  SYMLINK=$HOMEMC/temporary.$2.zoo
   if test -f $SYMLINK; then
-    SYMLINK=$HOME/.mc/temporary.$2.
+    SYMLINK=$HOMEMC/temporary.$2.
   fi
   if test -f $SYMLINK; then
     echo "Ugh. I did not expect this to happen. Cleanup your ~/.mc."
