diff -Nru mc-4.6.0.orig/doc/mc.1.in mc-4.6.0/doc/mc.1.in
--- mc-4.6.0.orig/doc/mc.1.in	Wed Feb  5 16:54:31 2003
+++ mc-4.6.0/doc/mc.1.in	Wed Oct  8 18:52:58 2003
@@ -3221,6 +3221,9 @@
 .IP
 The default system-wide extensions file.
 .PP
+.I $HOME_ETC/.mc/bindings
+.IP
+or
 .I ~/.mc/bindings
 .IP
 User's own extension, view configuration and edit configuration
@@ -3240,6 +3243,9 @@
 .\"Terminal databases"
 are loaded from mc.lib.
 .PP
+.I $HOME_ETC/.mc/ini
+.IP
+or
 .I ~/.mc/ini
 .IP
 User's own setup. If this file is present then the setup is loaded
@@ -3253,11 +3259,17 @@
 .IP
 This file contains the default system-wide applications menu.
 .PP
+.I $HOME_ETC/.mc/menu
+.IP
+or
 .I ~/.mc/menu
 .IP
 User's own application menu. If this file is present it is used instead
 of the system-wide applications menu.
 .PP
+.I $HOME_ETC/.mc/Tree
+.IP
+or
 .I ~/.mc/Tree
 .IP
 The directory list for the directory tree and tree view features.
diff -Nru mc-4.6.0.orig/doc/mcedit.1.in mc-4.6.0/doc/mcedit.1.in
--- mc-4.6.0.orig/doc/mcedit.1.in	Tue Jan 28 21:43:19 2003
+++ mc-4.6.0/doc/mcedit.1.in	Wed Oct  8 18:54:47 2003
@@ -472,16 +472,17 @@
 Global settings for the Midnight Commander.  Settings in this file
 affect all users, whether they have ~/.mc/ini or not.
 .PP
+.I $HOME_ETC/.mc/ini
+.IP
+or
 .I $HOME/.mc/ini
 .IP
 User's own setup.  If this file is present, the setup is loaded from
 here instead of the system-wide startup file.
 .PP
-.I $HOME/.mc/ini
+.I $HOME_ETC/.mc/cedit
 .IP
-User's own setup. If this file is present then the setup is loaded
-from here instead of the system-wide startup file.
-.PP
+or
 .I $HOME/.mc/cedit/
 .IP
 User's own temporary directory where block commands are processed
diff -Nru mc-4.6.0.orig/doc/mcview.1.in mc-4.6.0/doc/mcview.1.in
--- mc-4.6.0.orig/doc/mcview.1.in	Mon Jan 20 15:08:22 2003
+++ mc-4.6.0/doc/mcview.1.in	Wed Oct  8 18:55:28 2003
@@ -75,6 +75,9 @@
 Global settings for the Midnight Commander.  Settings in this file
 affect all users, whether they have ~/.mc/ini or not.
 .PP
+.I $HOME_ETC/.mc/ini
+.IP
+or
 .I $HOME/.mc/ini
 .IP
 User's own setup.  If this file is present, the setup is loaded from
diff -Nru mc-4.6.0.orig/edit/edit.c mc-4.6.0/edit/edit.c
--- mc-4.6.0.orig/edit/edit.c	Wed Oct  8 17:46:13 2003
+++ mc-4.6.0/edit/edit.c	Wed Oct  8 18:21:55 2003
@@ -2708,7 +2708,7 @@
     int nomark;
     struct stat status;
     long start_mark, end_mark;
-    char *block_file = catstrs (home_dir, BLOCK_FILE, 0);
+    char *block_file = catstrs (etc_dir, BLOCK_FILE, 0);
     int rc = 0;
 
     nomark = eval_marks (edit, &start_mark, &end_mark);
diff -Nru mc-4.6.0.orig/edit/editcmd.c mc-4.6.0/edit/editcmd.c
--- mc-4.6.0.orig/edit/editcmd.c	Thu Dec 19 14:01:34 2002
+++ mc-4.6.0/edit/editcmd.c	Wed Oct  8 18:21:55 2003
@@ -539,7 +539,7 @@
 {
     char *filename;
     int file;
-    filename = catstrs (home_dir, MACRO_FILE, 0);
+    filename = catstrs (etc_dir, MACRO_FILE, 0);
     if ((file = open (filename, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) == -1)
 	return 0;
     close (file);
@@ -575,7 +575,7 @@
     if (saved_macros_loaded)
 	if ((j = macro_exists (k)) < 0)
 	    return 0;
-    g = fopen (catstrs (home_dir, TEMP_FILE, 0), "w");
+    g = fopen (catstrs (etc_dir, TEMP_FILE, 0), "w");
     if (!g) {
 /* This heads the delete macro error dialog box */
 	edit_error_dialog (_(" Delete macro "),
@@ -609,7 +609,7 @@
     }
     fclose (f);
     fclose (g);
-    if (rename (catstrs (home_dir, TEMP_FILE, 0), catstrs (home_dir, MACRO_FILE, 0)) == -1) {
+    if (rename (catstrs (etc_dir, TEMP_FILE, 0), catstrs (etc_dir, MACRO_FILE, 0)) == -1) {
 /* This heads the delete macro error dialog box */
 	edit_error_dialog (_(" Delete macro "),
 	   get_sys_error (_(" Cannot overwrite macro file ")));
@@ -2115,7 +2115,7 @@
 /* copies a block to clipboard file */
 static int edit_save_block_to_clip_file (WEdit * edit, long start, long finish)
 {
-    return edit_save_block (edit, catstrs (home_dir, CLIP_FILE, 0), start, finish);
+    return edit_save_block (edit, catstrs (etc_dir, CLIP_FILE, 0), start, finish);
 }
 
 
@@ -2152,7 +2152,7 @@
 
 void edit_paste_from_X_buf_cmd (WEdit * edit)
 {
-    edit_insert_file (edit, catstrs (home_dir, CLIP_FILE, 0));
+    edit_insert_file (edit, catstrs (etc_dir, CLIP_FILE, 0));
 }
 
 
@@ -2204,7 +2204,7 @@
     if (eval_marks (edit, &start_mark, &end_mark))
 	return 1;
     exp =
-	edit_get_save_file (catstrs (home_dir, CLIP_FILE, 0),
+	edit_get_save_file (catstrs (etc_dir, CLIP_FILE, 0),
 			    _(" Save Block "));
     edit_push_action (edit, KEY_PRESS + edit->start_display);
     if (exp) {
@@ -2233,7 +2233,7 @@
 int
 edit_insert_file_cmd (WEdit *edit)
 {
-    char *exp = edit_get_load_file (catstrs (home_dir, CLIP_FILE, 0),
+    char *exp = edit_get_load_file (catstrs (etc_dir, CLIP_FILE, 0),
 				    _(" Insert File "));
     edit_push_action (edit, KEY_PRESS + edit->start_display);
     if (exp) {
@@ -2269,7 +2269,7 @@
 	edit_error_dialog (_(" Sort block "), _(" You must first highlight a block of text. "));
 	return 0;
     }
-    edit_save_block (edit, catstrs (home_dir, BLOCK_FILE, 0), start_mark, end_mark);
+    edit_save_block (edit, catstrs (etc_dir, BLOCK_FILE, 0), start_mark, end_mark);
 
     exp = old ? old : "";
 
@@ -2282,7 +2282,7 @@
 	g_free (old);
     old = exp;
 
-    e = system (catstrs (" sort ", exp, " ", home_dir, BLOCK_FILE, " > ", home_dir, TEMP_FILE, 0));
+    e = system (catstrs (" sort ", exp, " ", etc_dir, BLOCK_FILE, " > ", etc_dir, TEMP_FILE, 0));
     if (e) {
 	if (e == -1 || e == 127) {
 	    edit_error_dialog (_(" Sort "), 
@@ -2300,7 +2300,7 @@
 
     if (edit_block_delete_cmd (edit))
 	return 1;
-    edit_insert_file (edit, catstrs (home_dir, TEMP_FILE, 0));
+    edit_insert_file (edit, catstrs (etc_dir, TEMP_FILE, 0));
     return 0;
 }
 
@@ -2321,8 +2321,8 @@
     char *quoted_name = NULL;
 
     o = catstrs (mc_home, shell_cmd, 0);	/* original source script */
-    h = catstrs (home_dir, EDIT_DIR, shell_cmd, 0);	/* home script */
-    b = catstrs (home_dir, BLOCK_FILE, 0);	/* block file */
+    h = catstrs (etc_dir, EDIT_DIR, shell_cmd, 0);	/* home script */
+    b = catstrs (etc_dir, BLOCK_FILE, 0);	/* block file */
 
     if (!(script_home = fopen (h, "r"))) {
 	if (!(script_home = fopen (h, "w"))) {
@@ -2374,8 +2374,8 @@
 	 *   $3 - file where error messages should be put
 	 *        (for compatibility with old scripts).
 	 */
-	system (catstrs (" ", home_dir, EDIT_DIR, shell_cmd, " ", quoted_name,
-			 " ", home_dir, BLOCK_FILE " /dev/null", NULL));
+	system (catstrs (" ", etc_dir, EDIT_DIR, shell_cmd, " ", quoted_name,
+			 " ", etc_dir, BLOCK_FILE " /dev/null", NULL));
 
     } else {
 	/*
@@ -2383,7 +2383,7 @@
 	 * Arguments:
 	 *   $1 - name of the edited file.
 	 */
-	system (catstrs (" ", home_dir, EDIT_DIR, shell_cmd, " ",
+	system (catstrs (" ", etc_dir, EDIT_DIR, shell_cmd, " ",
 			 quoted_name, NULL));
     }
     g_free (quoted_name);
diff -Nru mc-4.6.0.orig/edit/editwidget.c mc-4.6.0/edit/editwidget.c
--- mc-4.6.0.orig/edit/editwidget.c	Wed Dec 25 23:53:07 2002
+++ mc-4.6.0/edit/editwidget.c	Wed Oct  8 18:21:55 2003
@@ -166,7 +166,7 @@
 		sizeof (int));
     }
     if (!made_directory) {
-	mkdir (catstrs (home_dir, EDIT_DIR, 0), 0700);
+	mkdir (catstrs (etc_dir, EDIT_DIR, 0), 0700);
 	made_directory = 1;
     }
 
diff -Nru mc-4.6.0.orig/edit/syntax.c mc-4.6.0/edit/syntax.c
--- mc-4.6.0.orig/edit/syntax.c	Sun Dec 15 19:55:53 2002
+++ mc-4.6.0/edit/syntax.c	Wed Oct  8 18:21:55 2003
@@ -562,7 +562,7 @@
 	return fopen (filename, "r");
 
     g_free (error_file_name);
-    error_file_name = g_strconcat (home_dir, EDIT_DIR PATH_SEP_STR,
+    error_file_name = g_strconcat (etc_dir, EDIT_DIR PATH_SEP_STR,
 				   filename, NULL);
     f = fopen (error_file_name, "r");
     if (f)
@@ -974,7 +974,7 @@
 	if (!*edit->filename && !type)
 	    return;
     }
-    f = catstrs (home_dir, SYNTAX_FILE, 0);
+    f = catstrs (etc_dir, SYNTAX_FILE, 0);
     r = edit_read_syntax_file (edit, names, f, edit ? edit->filename : 0,
 			       get_first_editor_line (edit), type);
     if (r == -1) {
diff -Nru mc-4.6.0.orig/man/pl/man1/mc.1 mc-4.6.0/man/pl/man1/mc.1
--- mc-4.6.0.orig/man/pl/man1/mc.1	Thu Jun 14 10:43:13 2001
+++ mc-4.6.0/man/pl/man1/mc.1	Wed Oct  8 18:43:54 2003
@@ -2758,10 +2758,14 @@
 .IP
 Standardowy plik rozszerzeñ plików.
 .PP
+$HOME_ETC/.mc/ext
+.IP albo
+.PP
 ~/.mc/ext
 .IP
 W³asny plik u¿ytkownika, konfiguruje podgl±d i edycje plików. Ma wy¿szy
-priorytet ni¿ plik systemowy.
+priorytet ni¿ plik systemowy. Gdy ustawiona jest zmienna $HOME_ETC plik
+$HOME_ETC/.mc/ext jest u¿ywany w miejsce ~/.mc/ext.
 .PP
 /usr/lib/mc/mc.ini
 .IP
@@ -2774,10 +2778,15 @@
 uwzglêdniane przez wszystkie sesje Midnight Commandera, u¿yteczne do
 definiowania ogólnosystemowych ustawieñ terminali.
 .PP
+$HOME_ETC/.mc/ini
+.IP
+albo
+.PP
 ~/.mc/ini
 .IP
 W³asny setup u¿ytkownika. Je¶li ten plik jest dostêpny, jest ³adowany
-zamiast pliku globalnego.
+zamiast pliku globalnego. Gdy ustawiona jest zmienna $HOME_ETC plik
+$HOME_ETC/.mc/ini jest u¿ywany zamiast ~/.mc/ini.
 .PP
 /usr/lib/mc/mc.hint
 .IP
@@ -2787,10 +2796,19 @@
 .IP
 Ten plik zawiera informacje o ogólnosystemowych aplikacjach w menu.
 .PP
+$HOME_ETC/.mc/menu
+.IP
+albo
+.PP
 ~/.mc/menu
 .IP
 W³asny plik menu u¿ytkownika. Je¶li ten plik jest obecny jest u¿ywany
-zamiast pliku globalnego.
+zamiast pliku globalnego. Gdy ustawiona jest zmienna $HOME_ETC plik
+$HOME_ETC/.mc/menu jest u¿ywany zamiast ~/.mc/menu.
+.PP
+$HOME_ETC
+.IP
+albo
 .PP
 ~/.mc/tree
 .IP
@@ -2799,7 +2817,8 @@
 Linie zaczynaj±ce siê od numeru maj± tyle znaków ile poprzedni katalog.
 Je¶li chcesz mo¿esz stworzyæ plik u¿ywaj±c komendy "find / -type d
 -print | sort > ~/.mc/tree". Normalnie nie ma sensu tego czyniæ, gdy¿
-Midnight Commander robi to sam za ciebie.
+Midnight Commander robi to sam za ciebie. Gdy ustawiona jest zmienna
+$HOME_ETC plik $HOME_ETC/.mc/tree jest u¿ywany zamiast ~/.mc/tree
 .PP
 \&./.mc.menu
 .IP
diff -Nru mc-4.6.0.orig/man/pl/man1/mcedit.1 mc-4.6.0/man/pl/man1/mcedit.1
--- mc-4.6.0.orig/man/pl/man1/mcedit.1	Sat Sep 30 19:08:37 2000
+++ mc-4.6.0/man/pl/man1/mcedit.1	Wed Oct  8 18:48:17 2003
@@ -418,15 +418,23 @@
 dla ka¿dej sesji Midnight Commandera. Przydatny do zdefiniowania ustawieñ
 terminala obowi±zuj±cych dla ca³ego systemu (site-global).
 .PP
+.I $HOME_ETC/.mc.ini
+.IP
+albo
 .I $HOME/.mc.ini
 .IP
 Prywatna konfiguracja u¿ytkownika. Je¶li istnieje, to konfiguracja odczytywana
 jest z tego w³a¶nie pliku, zamiast z ogólnosystemowego pliku startowego.
+Zmienna ¶rodowiskowa $HOME_ETC ma pierwszeñstwo przed $HOME.
 .PP
+.I $HOME_ETC/.cedit
+.IP
+albo
 .I $HOME/.cedit/
 .IP
 Tymczasowy katalog u¿ytkownika. Przetwarzane i zapisywane s± w nim polecenia
-operuj±ce na blokach tekstu.
+operuj±ce na blokach tekstu. Zmienna ¶rodowiskowa $HOME_ETC ma pierwszeñstwo
+przed $HOME.
 .SH LICENCJA
 Niniejszy program jest rozpowszechniany na warunkach licencji
 GNU General Public Licence opublikowanych przez Free Software Foundation.
diff -Nru mc-4.6.0.orig/slang/sltermin.c mc-4.6.0/slang/sltermin.c
--- mc-4.6.0.orig/slang/sltermin.c	Mon Oct  7 13:08:16 2002
+++ mc-4.6.0/slang/sltermin.c	Wed Oct  8 18:21:55 2003
@@ -265,7 +265,7 @@
    /* If we are on a termcap based system, use termcap */
    if (0 == tcap_getent (term, ti)) return ti;
 
-   if (NULL != (home = getenv ("HOME")))
+   if (NULL != (home = getenv("HOME_ETC") ? getenv("HOME_ETC") : getenv ("HOME")))
      {
 	strncpy (home_ti, home, sizeof (home_ti) - 11);
 	home_ti [sizeof(home_ti) - 11] = 0;
diff -Nru mc-4.6.0.orig/src/cmd.c mc-4.6.0/src/cmd.c
--- mc-4.6.0.orig/src/cmd.c	Wed Oct  8 17:46:13 2003
+++ mc-4.6.0/src/cmd.c	Wed Oct  8 18:21:55 2003
@@ -643,7 +643,7 @@
     extdir = concat_dir_and_file (mc_home, MC_LIB_EXT);
 
     if (dir == 0){
-	buffer = concat_dir_and_file (home_dir, MC_USER_EXT);
+	buffer = concat_dir_and_file (etc_dir, MC_USER_EXT);
 	check_for_default (extdir, buffer);
 	do_edit (buffer);
 	g_free (buffer);
@@ -679,7 +679,7 @@
 	    break;
 
 	case 1:
-	    buffer = concat_dir_and_file (home_dir, where ? CEDIT_HOME_MENU : MC_HOME_MENU);
+	    buffer = concat_dir_and_file (etc_dir, where ? CEDIT_HOME_MENU : MC_HOME_MENU);
 	    check_for_default (menufile, buffer);
 	    break;
 	
@@ -746,7 +746,7 @@
     extdir = concat_dir_and_file (mc_home, "syntax" PATH_SEP_STR "Syntax");
 
     if (dir == 0) {
-	buffer = concat_dir_and_file (home_dir, SYNTAX_FILE);
+	buffer = concat_dir_and_file (etc_dir, SYNTAX_FILE);
 	check_for_default (extdir, buffer);
 	do_edit (buffer);
 	g_free (buffer);
diff -Nru mc-4.6.0.orig/src/ext.c mc-4.6.0/src/ext.c
--- mc-4.6.0.orig/src/ext.c	Thu Nov 14 08:25:19 2002
+++ mc-4.6.0/src/ext.c	Wed Oct  8 18:21:55 2003
@@ -497,7 +497,7 @@
 	int mc_user_ext = 1;
 	int home_error = 0;
 
-	extension_file = concat_dir_and_file (home_dir, MC_USER_EXT);
+	extension_file = concat_dir_and_file (etc_dir, MC_USER_EXT);
 	if (!exist_file (extension_file)) {
 	    g_free (extension_file);
 	  check_stock_mc_ext:
diff -Nru mc-4.6.0.orig/src/global.h mc-4.6.0/src/global.h
--- mc-4.6.0.orig/src/global.h	Mon Dec 16 01:41:06 2002
+++ mc-4.6.0/src/global.h	Wed Oct  8 18:21:55 2003
@@ -141,6 +141,7 @@
 #include "textconf.h"
 #include "../vfs/vfs.h"
 
+extern char *etc_dir;
 extern char *home_dir;
 
 #ifdef min
diff -Nru mc-4.6.0.orig/src/hotlist.c mc-4.6.0/src/hotlist.c
--- mc-4.6.0.orig/src/hotlist.c	Sat Dec 21 09:43:15 2002
+++ mc-4.6.0/src/hotlist.c	Wed Oct  8 18:21:55 2003
@@ -1363,7 +1363,7 @@
     }
 
     if (!hotlist_file_name)
-	hotlist_file_name = concat_dir_and_file (home_dir, HOTLIST_FILENAME);
+	hotlist_file_name = concat_dir_and_file (etc_dir, HOTLIST_FILENAME);
     
     hotlist	       = new_hotlist ();
     hotlist->type      = HL_TYPE_GROUP;
diff -Nru mc-4.6.0.orig/src/main.c mc-4.6.0/src/main.c
--- mc-4.6.0.orig/src/main.c	Wed Oct  8 17:46:13 2003
+++ mc-4.6.0/src/main.c	Wed Oct  8 18:21:55 2003
@@ -248,6 +248,9 @@
 char *program_name;
 #endif
 
+/* Configuration directory */
+char *etc_dir;
+
 /* The home directory */
 char *home_dir;
 
@@ -2395,7 +2398,7 @@
 do_mc_filename_rename (char *mc_dir, char *o_name, char *n_name)
 {
     char *full_o_name = concat_dir_and_file (home_dir, o_name);
-    char *full_n_name = g_strconcat (home_dir, MC_BASE, n_name, NULL);
+    char *full_n_name = g_strconcat (etc_dir, MC_BASE, n_name, NULL);
     int move;
 
     move = 0 == rename (full_o_name, full_n_name);
@@ -2409,7 +2412,7 @@
 {
     struct stat s;
     int move = 0;
-    char *mc_dir = concat_dir_and_file (home_dir, ".mc");
+    char *mc_dir = concat_dir_and_file (etc_dir, ".mc");
 
     if (stat (mc_dir, &s) && (errno == ENOENT)
 	&& (mkdir (mc_dir, 0777) != -1)) {
@@ -2448,11 +2451,13 @@
     OS_Setup ();
 
     /* This variable is used by the subshell */
-    home_dir = getenv ("HOME");
+    etc_dir = getenv ("HOME_ETC");
+		home_dir = getenv ("HOME");
     if (!home_dir) {
 	/* mc_home was computed by OS_Setup */
 	home_dir = mc_home;
     }
+    if (!etc_dir) etc_dir = home_dir;
 
     vfs_init ();
 
diff -Nru mc-4.6.0.orig/src/poptconfig.c mc-4.6.0/src/poptconfig.c
--- mc-4.6.0.orig/src/poptconfig.c	Thu Oct 24 12:02:40 2002
+++ mc-4.6.0/src/poptconfig.c	Wed Oct  8 18:21:55 2003
@@ -134,8 +134,10 @@
     rc = poptReadConfigFile(con, "/etc/popt");
     if (rc) return rc;
     if (getuid() != geteuid()) return 0;
+    home = getenv("HOME_ETC");
+		if (!home) home = getenv("HOME");
 
-    if ((home = getenv("HOME"))) {
+    if (home) {
 	fn = alloca(strlen(home) + 20);
 	strcpy(fn, home);
 	strcat(fn, "/.popt");
diff -Nru mc-4.6.0.orig/src/setup.c mc-4.6.0/src/setup.c
--- mc-4.6.0.orig/src/setup.c	Wed Oct  8 17:46:13 2003
+++ mc-4.6.0/src/setup.c	Wed Oct  8 18:21:55 2003
@@ -269,7 +269,7 @@
     int  i;
     char buffer [BUF_TINY];
 
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = concat_dir_and_file (etc_dir, PROFILE_NAME);
 
     /* Save integer options */
     for (i = 0; layout [i].opt_name; i++){
@@ -286,7 +286,7 @@
     char *profile;
     int  i;
 
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = concat_dir_and_file (etc_dir, PROFILE_NAME);
 
     /* Save integer options */
     for (i = 0; options [i].opt_name; i++)
@@ -329,7 +329,7 @@
     char *profile;
 
     saving_setup = 1;
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = concat_dir_and_file (etc_dir, PROFILE_NAME);
 
     save_configure ();
 
@@ -467,7 +467,7 @@
     if (profile_name)
 	    return profile_name;
 
-    profile = concat_dir_and_file (home_dir, PROFILE_NAME);
+    profile = concat_dir_and_file (etc_dir, PROFILE_NAME);
     if (!exist_file (profile)){
 	inifile = concat_dir_and_file (mc_home, "mc.ini");
 	if (exist_file (inifile)){
diff -Nru mc-4.6.0.orig/src/treestore.c mc-4.6.0/src/treestore.c
--- mc-4.6.0.orig/src/treestore.c	Mon Sep 23 08:43:23 2002
+++ mc-4.6.0/src/treestore.c	Wed Oct  8 18:21:55 2003
@@ -272,7 +272,7 @@
     char *name;
     int retval;
 
-    name = concat_dir_and_file(home_dir, MC_TREE);
+    name = concat_dir_and_file(etc_dir, MC_TREE);
     retval = tree_store_load_from(name);
     g_free(name);
 
@@ -377,7 +377,7 @@
     char *name;
     int retval;
 
-    tmp = concat_dir_and_file(home_dir, MC_TREE_TMP);
+    tmp = concat_dir_and_file(etc_dir, MC_TREE_TMP);
     retval = tree_store_save_to(tmp);
 
     if (retval) {
@@ -385,7 +385,7 @@
 	return retval;
     }
 
-    name = concat_dir_and_file(home_dir, MC_TREE);
+    name = concat_dir_and_file(etc_dir, MC_TREE);
     retval = rename(tmp, name);
 
     g_free(tmp);
diff -Nru mc-4.6.0.orig/src/user.c mc-4.6.0/src/user.c
--- mc-4.6.0.orig/src/user.c	Fri Nov 29 04:03:53 2002
+++ mc-4.6.0/src/user.c	Wed Oct  8 18:21:55 2003
@@ -219,7 +219,7 @@
     case 'k':			/* block file name */
     case 'b':			/* block file name / strip extension */  {
 	    if (edit_widget) {
-		char *file = g_strconcat (home_dir, BLOCK_FILE, NULL);
+		char *file = g_strconcat (etc_dir, BLOCK_FILE, NULL);
 		fname = (*quote_func) (file, 0);
 		g_free (file);
 		return fname;
@@ -688,7 +688,7 @@
     if (!exist_file (menu) || !menu_file_own (menu)){
 	g_free (menu);
         menu = concat_dir_and_file \
-                            (home_dir, edit_widget ? CEDIT_HOME_MENU : MC_HOME_MENU);
+                            (etc_dir, edit_widget ? CEDIT_HOME_MENU : MC_HOME_MENU);
 	if (!exist_file (menu)){
 	    g_free (menu);
 	    menu = concat_dir_and_file \
diff -Nru mc-4.6.0.orig/src/util.c mc-4.6.0/src/util.c
--- mc-4.6.0.orig/src/util.c	Tue Jan 28 23:58:23 2003
+++ mc-4.6.0/src/util.c	Wed Oct  8 18:21:55 2003
@@ -1266,7 +1266,7 @@
     *column = 0;
 
     /* open file with positions */
-    fn = concat_dir_and_file (home_dir, MC_FILEPOS);
+    fn = concat_dir_and_file (etc_dir, MC_FILEPOS);
     f = fopen (fn, "r");
     g_free (fn);
     if (!f)
@@ -1309,8 +1309,8 @@
 
     len = strlen (filename);
 
-    tmp = concat_dir_and_file (home_dir, MC_FILEPOS_TMP);
-    fn = concat_dir_and_file (home_dir, MC_FILEPOS);
+    tmp = concat_dir_and_file (etc_dir, MC_FILEPOS_TMP);
+    fn = concat_dir_and_file (etc_dir, MC_FILEPOS);
 
     /* open temporary file */
     t = fopen (tmp, "w");
diff -Nru mc-4.6.0.orig/src/widget.c mc-4.6.0/src/widget.c
--- mc-4.6.0.orig/src/widget.c	Thu Dec 26 00:15:48 2002
+++ mc-4.6.0/src/widget.c	Wed Oct  8 18:30:39 2003
@@ -815,7 +815,7 @@
 	return 0;
     if (!*input_name)
 	return 0;
-    profile = concat_dir_and_file (home_dir, HISTORY_FILE_NAME);
+    profile = concat_dir_and_file (etc_dir, HISTORY_FILE_NAME);
     for (i = 0;; i++) {
 	char key_name[BUF_TINY];
 	char this_entry[BUF_LARGE];
@@ -852,7 +852,7 @@
     if (!num_history_items_recorded)	/* this is how to disable */
 	return;
 
-    profile = concat_dir_and_file (home_dir, HISTORY_FILE_NAME);
+    profile = concat_dir_and_file (etc_dir, HISTORY_FILE_NAME);
 
     if ((i = open (profile, O_CREAT | O_EXCL, S_IRUSR | S_IWUSR)) != -1)
 	close (i);
diff -Nru mc-4.6.0.orig/vfs/ftpfs.c mc-4.6.0/vfs/ftpfs.c
--- mc-4.6.0.orig/vfs/ftpfs.c	Thu Dec 26 03:21:43 2002
+++ mc-4.6.0/vfs/ftpfs.c	Wed Oct  8 18:21:55 2003
@@ -1954,7 +1954,7 @@
     }
 
     /* Load current .netrc */
-    netrcname = concat_dir_and_file (home_dir, ".netrc");
+    netrcname = concat_dir_and_file (etc_dir, ".netrc");
     netrcp = netrc = load_file (netrcname);
     if (netrc == NULL) {
 	g_free (netrcname);
