--- mc-4.5.55/src/main.c~	Sun Sep 16 15:29:00 2001
+++ mc-4.5.55/src/main.c	Sun Sep 16 15:30:19 2001
@@ -2373,7 +2373,7 @@
 	fprintf (stderr, _("The TERM environment variable is unset!\n"));
 	exit (1);
     }
-    if (force_xterm || (strncmp (termvalue, "xterm", 5) == 0 || strcmp (termvalue, "dtterm") == 0)){
+    if (force_xterm || (strncmp (termvalue, "xterm", 5) == 0 || strcmp (termvalue, "dtterm") == 0 || strncmp (termvalue, "rxvt", 4) == 0)){
 	use_mouse_p = XTERM_MOUSE;
 	xterm_flag = 1;
 #    ifdef SET_TITLE
