diff -urN mc-4.6.0.orig/lib/mc.ext.in mc-4.6.0/lib/mc.ext.in
--- mc-4.6.0.orig/lib/mc.ext.in	2004-03-26 10:31:23.330492768 +0100
+++ mc-4.6.0/lib/mc.ext.in	2004-03-26 10:32:01.078754160 +0100
@@ -300,7 +300,7 @@
 
 # html
 regex/\.([Hh]tml?|HTML?)$
-	Open=if test -n "@X11_WWW@" && test -n "$DISPLAY"; then (@X11_WWW@ file://%d/%p &) >/dev/null 2>&1; else links %f 2>/dev/null || lynx -force_html %f; fi
+	Open=if test -n "gnome-moz-remote" && test -n "$DISPLAY"; then (gnome-moz-remote file://%d/%p &) >/dev/null 2>&1; else elinks %f 2>/dev/null || lynx -force_html %f; fi
 	View=%view{ascii} lynx -dump -force_html %f
 
 # StarOffice 5.2
@@ -442,34 +442,49 @@
 # cpio
 shell/.cpio.Z
 	Open=%cd %p#ucpio
-	View=%view{ascii} gzip -dc '%f' | cpio -itv 2>/dev/null
+	View=%view{ascii} gzip -dc %f | cpio -itv 2>/dev/null
 
 shell/.cpio.gz
 	Open=%cd %p#ucpio
-	View=%view{ascii} gzip -dc '%f' | cpio -itv 2>/dev/null
+	View=%view{ascii} gzip -dc %f | cpio -itv 2>/dev/null
 
 shell/.cpio
 	Open=%cd %p#ucpio
-	View=%view{ascii} cpio -itv <'%f' 2>/dev/null
+	View=%view{ascii} cpio -itv < %f 2>/dev/null
 
 # ls-lR
-regex/(^|\.)ls-?lR(\.g?z|Z|bz2)?$
+regex/(^|\.)ls-?lR(\.gz|Z|bz2)$
 	Open=%cd %p#lslR
 
 # patch
-regex/\.(diff|patch)(\.(bz2|gz|Z))?$
+regex/\.(diff|patch)(\.bz2)$
+	Open=%cd %p#patchfs
+	View=%view{ascii} bzip2 -dc %f 2>/dev/null
+
+regex/\.(diff|patch)(\.(gz|Z))$
+	Open=%cd %p#patchfs
+	View=%view{ascii} gzip -dc %f 2>/dev/null
+
+regex/\.(diff|patch)$
         Open=%cd %p#patchfs
+	View=%view{ascii} /bin/cat %f 2>/dev/null
 
 # gzip
 type/^gzip
 	Open=gzip -dc %f | %var{PAGER:more}
 	View=%view{ascii} gzip -dc %f 2>/dev/null
 
+regex/\.(gz|Z)?$
+	View=%view{ascii} gzip -dc %f 2>/dev/null
+
 # bzip2
 type/^bzip2
 	Open=bzip2 -dc %f | %var{PAGER:more}
 	View=%view{ascii} bzip2 -dc %f 2>/dev/null
 
+regex/\.bz2?$
+	View=%view{ascii} bzip2 -dc %f 2>/dev/null
+
 # bzip
 type/^bzip
 	Open=bzip -dc %f | %var{PAGER:more}
