diff -urN mc-4.6.0/src/screen.c mc-4.6.0.new/src/screen.c
--- mc-4.6.0/src/screen.c	2003-01-28 23:58:22.000000000 +0100
+++ mc-4.6.0.new/src/screen.c	2003-09-27 08:33:09.000000000 +0200
@@ -2009,6 +2009,28 @@
 }
 
 static void
+chdir_other_panel_old (WPanel *panel)
+{
+    char *new_dir;
+
+    if (get_other_type () != view_listing)
+	return;
+
+    if (!S_ISDIR (panel->dir.list [panel->selected].buf.st_mode))
+	new_dir = concat_dir_and_file (panel->cwd, "..");
+    else
+	new_dir = concat_dir_and_file (panel->cwd, panel->dir.list [panel->selected].fname);
+
+    change_panel ();
+    do_cd (new_dir, cd_exact);
+    change_panel ();
+
+    move_down (panel);
+
+    g_free (new_dir);
+}
+
+static void
 chdir_to_readlink (WPanel *panel)
 {
     char *new_dir;
@@ -2090,6 +2112,7 @@
     { ALT('s'),   start_search },	/* M-s not like emacs */
     { XCTRL('t'), mark_file },
     { ALT('o'),   chdir_other_panel },
+    { ALT('i'),   chdir_other_panel_old },
     { ALT('l'),   chdir_to_readlink },
     { ALT('H'),   directory_history_list },
     { KEY_F(13),  view_simple_cmd },
