--- mc-4.6.0/lib/mc.ext.in~	Fri Apr 25 00:12:46 2003
+++ mc-4.6.0/lib/mc.ext.in	Fri Apr 25 00:39:42 2003
@@ -427,7 +427,6 @@
 	Open=%cd %p#uesp
 	View=%view{ascii} unesp v %f
 	Extract=unesp x %f
-	Icon=compressed.xpm
 
 # ha
 regex/\.(ha|HA|Ha)$
@@ -457,19 +456,30 @@
 	Open=%cd %p#lslR
 
 # patch
-regex/\.(diff|patch)(\.(bz2|gz|Z))?$
-        Open=%cd %p#patchfs
+regex/\.(diff|patch).bz2?$
+	Open=%cd %p#patchfs
+	View=%view{ascii} bzip2 -dc %f 2>/dev/null
+
+regex/\.(diff|patch)\.(gz|Z)?$
+	Open=%cd %p#patchfs
+	View=%view{ascii} gzip -dc %f 2>/dev/null
 
 # gzip
 type/^gzip
 	Open=gzip -dc %f | %var{PAGER:more}
 	View=%view{ascii} gzip -dc %f 2>/dev/null
 
+regex/\.(gz|Z)?$
+	View=%view{ascii} gzip -dc %f 2>/dev/null
+
 # bzip2
 type/^bzip2
 	Open=bzip2 -dc %f | %var{PAGER:more}
 	View=%view{ascii} bzip2 -dc %f 2>/dev/null
 
+regex/\.bz2?$
+	View=%view{ascii} bzip2 -dc %f 2>/dev/null
+
 # bzip
 type/^bzip
 	Open=bzip -dc %f | %var{PAGER:more}
