--- mc-4.5.36/gnome/mc.keys.in.in.mimekeys	Wed Jul  7 16:39:38 1999
+++ mc-4.5.36/gnome/mc.keys.in.in	Wed Jul  7 16:40:45 1999
@@ -77,6 +77,11 @@
 	edit=applix %f
 	view=applix %f
 
+application/x-applix-word
+	open=applix %f
+	edit=applix %f
+	view=applix %f
+
 application/x-compressed-tar
 	fm-open=%cd %p#utar
 	fm-view=%view{ascii} gzip -dc %f 2>/dev/null | tar tvvf -
@@ -285,6 +290,10 @@
 	icon-filename=@icondir@/gnome-application-x-cc-source.png
 
 text/x-comma-separated-values:
+	open=gnumeric %f
+	view=gnumeric %f
+
+text/tab-separated-values
 	open=gnumeric %f
 	view=gnumeric %f
 
