--- ./lib/mc.ext.in.org	2005-07-27 20:52:12.000000000 +0200
+++ ./lib/mc.ext.in	2005-07-27 20:59:04.000000000 +0200
@@ -164,23 +164,32 @@
 # cpio
 shell/.cpio.Z
 	Open=%cd %p#ucpio
-	View=%view{ascii} gzip -dc '%f' | cpio -itv 2>/dev/null
+	View=%view{ascii} gzip -dc %f | cpio -itv 2>/dev/null
 
 shell/.cpio.gz
 	Open=%cd %p#ucpio
-	View=%view{ascii} gzip -dc '%f' | cpio -itv 2>/dev/null
+	View=%view{ascii} gzip -dc %f | cpio -itv 2>/dev/null
 
 shell/.cpio
 	Open=%cd %p#ucpio
-	View=%view{ascii} cpio -itv <'%f' 2>/dev/null
+	View=%view{ascii} cpio -itv < %f 2>/dev/null
 
 # ls-lR
-regex/(^|\.)ls-?lR(\.g?z|Z|bz2)?$
+regex/(^|\.)ls-?lR(\.gz|Z|bz2)$
 	Open=%cd %p#lslR
 
 # patch
-regex/\.(diff|patch)(\.(bz2|gz|Z))?$
-        Open=%cd %p#patchfs
+regex/\.(diff|patch)(\.bz2)$
+	Open=%cd %p#patchfs
+	View=%view{ascii} bzip2 -dc %f 2>/dev/null
+
+regex/\.(diff|patch)(\.(gz|Z))$
+	Open=%cd %p#patchfs
+	View=%view{ascii} gzip -dc %f 2>/dev/null
+
+regex/\.(diff|patch)$
+	Open=%cd %p#patchfs
+	View=%view{ascii} /bin/cat %f 2>/dev/null
 
 # ar library
 regex/\.s?a$
@@ -432,6 +441,7 @@
 # html
 regex/\.([hH][tT][mM][lL]?)$
 	Open=(if test -n "@X11_WWW@" && test -n "$DISPLAY"; then (@X11_WWW@ file://%d/%p &) 1>&2; else links %f || lynx -force_html %f || ${PAGER:-more} %f; fi) 2>/dev/null
+	 Open=(if test -n "gnome-moz-remote" && test -n "$DISPLAY"; then (gnome-moz-remote file://%d/%p &) >/dev/null 2>&1; else elinks %f 2>/dev/null || lynx -force_html %f || ${PAGER:-more} %f; fi) 2>/dev/null
 	View=%view{ascii} lynx -dump -force_html %f
 
 # StarOffice 5.2
@@ -526,11 +536,17 @@
 	Open=gzip -dc %f | %var{PAGER:more}
 	View=%view{ascii} gzip -dc %f 2>/dev/null
 
+regex/\.(gz|Z)?$
+	View=%view{ascii} gzip -dc %f 2>/dev/null
+
 # bzip2
 type/^bzip2
 	Open=bzip2 -dc %f | %var{PAGER:more}
 	View=%view{ascii} bzip2 -dc %f 2>/dev/null
 
+regex/\.bz2?$
+	View=%view{ascii} bzip2 -dc %f 2>/dev/null
+
 # bzip
 type/^bzip
 	Open=bzip -dc %f | %var{PAGER:more}
