--- mc-4.7.0-pre2/m4.include/mc-vfs.m4~	2009-08-04 23:36:33.000000000 +0200
+++ mc-4.7.0-pre2/m4.include/mc-vfs.m4	2009-09-06 18:44:17.955791270 +0200
@@ -40,7 +40,7 @@
 	dnl FIXME: network checks should probably be in their own macro.
 	AC_SEARCH_LIBS(socket, [xnet bsd socket inet], [have_socket=yes])
 	if test x"$have_socket" = xyes; then
-	    AC_SEARCH_LIBS(gethostbyname, [bsd socket inet netinet])
+	    AC_SEARCH_LIBS(gethostbyname, [bsd socket inet netinet nsl])
 	    AC_CHECK_MEMBERS([struct linger.l_linger], , , [
 #include <sys/types.h>
 #include <sys/socket.h>
