--- mc-4.6.0/lib/Makefile.am~	Thu Apr 24 09:57:02 2003
+++ mc-4.6.0/lib/Makefile.am	Thu Apr 24 09:57:02 2003
@@ -1,5 +1,4 @@
 suppbindir = $(pkgdatadir)/bin
-tidir      = $(pkgdatadir)/term
 
 if CHARSET
 LIBFILES_ADD = mc.charsets
@@ -22,7 +21,6 @@
 SCRIPTS_OUT = mc.csh mc.sh mc-wrapper.csh mc-wrapper.sh
 
 suppbin_SCRIPTS = $(SCRIPTS_OUT)
-ti_DATA = README.xterm linux.ti xterm.ad xterm.ti ansi.ti vt100.ti xterm.tcap
 
 pkgdata_DATA = $(LIBFILES_CONST) $(LIBFILES_ADD) $(LIBFILES_OUT)
 
@@ -35,6 +33,7 @@
 	$(ti_DATA) \
 	$(noinst_DATA) \
 	$(noinst_SCRIPTS) \
+	README.xterm linux.ti xterm.ad xterm.ti ansi.ti vt100.ti xterm.tcap \
 	mc.charsets
 
 mc.csh: $(srcdir)/mc.csh.in
