--- mc-4.6.1-pre1/configure.ac.orig	2004-07-23 17:51:54.373386000 +0000
+++ mc-4.6.1-pre1/configure.ac	2004-07-23 18:42:08.714170339 +0000
@@ -291,9 +291,6 @@
 dnl Network related functions
 dnl
 
-AC_CHECK_LIB(nsl, t_accept)
-AC_CHECK_LIB(socket, socket)
-
 have_socket=no
 AC_CHECK_FUNCS(socket, have_socket=yes)
 if test $have_socket = no; then
@@ -311,7 +308,7 @@
 AC_CHECK_FUNC(gethostbyname, [have_gethostbyname=yes])
 if test $have_gethostbyname = no; then
   # gethostbyname is not in the default libraries.  See if it's in some other.
-  for lib in bsd socket inet; do
+  for lib in bsd socket inet nsl; do
     AC_CHECK_LIB([$lib], [gethostbyname],
 		 [LIBS="$LIBS -l$lib"; have_gethostbyname=yes; break])
   done
--- mc-4.6.1-pre1/acinclude.m4.orig	2003-11-28 06:18:32.000000000 +0000
+++ mc-4.6.1-pre1/acinclude.m4	2004-07-23 18:50:44.777640579 +0000
@@ -144,10 +144,9 @@
 
   if test "x$enable_netcode" != xno; then
     dnl FIXME: network checks should probably be in their own macro.
-    AC_CHECK_LIB(nsl, main)
     AC_SEARCH_LIBS(socket, [xnet bsd socket inet], [have_socket=yes])
     if test x$have_socket = xyes; then
-      AC_SEARCH_LIBS(gethostbyname, [bsd socket inet netinet])
+      AC_SEARCH_LIBS(gethostbyname, [bsd socket inet netinet nsl])
       AC_CHECK_MEMBERS([struct linger.l_linger], , , [
 #include <sys/types.h>
 #include <sys/socket.h>
