--- mc-4.5.54/lib/mc.sh.in.chmou	Tue Oct  3 21:58:53 2000
+++ mc-4.5.54/lib/mc.sh.in	Thu Mar 22 15:11:11 2001
@@ -9,4 +9,4 @@
         unset MC;
 }
 
-export -f mc
+export mc
diff -Nur mc-4.5.54.orig/vfs/extfs/a mc-4.5.54/vfs/extfs/a
--- mc-4.5.54.orig/vfs/extfs/a	Fri Sep 15 21:10:07 2000
+++ mc-4.5.54/vfs/extfs/a	Sat Aug  4 16:13:38 2001
@@ -26,29 +26,29 @@
   /mkdir/ && do {
     shift; shift;
     exit 1 if scalar(@ARGV) != 1;
-    system("$mmd $disk:/$ARGV[0] >& /dev/null");
+    system("$mmd $disk:/$ARGV[0] >/dev/null 2>&1");
     exit 0; };
   /rmdir/ && do {
     shift; shift;
     exit 1 if scalar(@ARGV) != 1;
-    system("$mrd $disk:/$ARGV[0] >& /dev/null");
+    system("$mrd $disk:/$ARGV[0] >/dev/null 2>&1");
     exit 0; };
   /rm/ && do {
     shift; shift;
     exit 1 if scalar(@ARGV) != 1;
-    system("$mdel $disk:/$ARGV[0] >& /dev/null");
+    system("$mdel $disk:/$ARGV[0] >/dev/null 2>&1");
     exit 0; };
   /copyout/ && do {
     shift; shift;
     exit 1 if scalar(@ARGV) != 2;
     ( $src, $dest ) = @ARGV;
-    system("$mcopy $disk:/$src $dest >& /dev/null");
+    system("$mcopy $disk:/$src $dest >/dev/null 2>&1");
     exit 0; };
   /copyin/ && do {
     shift; shift;
     exit 1 if scalar(@ARGV) != 2;
     ( $dest, $src ) = @ARGV;
-    system("$mcopy $src $disk:/$dest >& /dev/null");
+    system("$mcopy $src $disk:/$dest >/dev/null 2>&1");
     exit 0; };
   /.*/ && do {                               # an unfamiliar command
     exit 1; };
diff -Nur mc-4.5.54.orig/vfs/extfs/ulha.in mc-4.5.54/vfs/extfs/ulha.in
--- mc-4.5.54.orig/vfs/extfs/ulha.in	Sat Apr 10 01:18:56 1999
+++ mc-4.5.54/vfs/extfs/ulha.in	Sat Aug  4 16:12:54 2001
@@ -34,7 +34,7 @@
 # Define your awk
 AWK=@AWK@
 
-if ls -de . >& /dev/null;
+if ls -de . >/dev/null 2>&1 ;
 then
     LS_COMMAND="ls -le"
 else
