diff -urN mc-4.6.0-orig/configure.in mc-4.6.0/configure.in
--- mc-4.6.0-orig/configure.in	2004-02-06 21:06:20.000000000 +0100
+++ mc-4.6.0/configure.in	2004-02-06 21:09:06.553523944 +0100
@@ -571,24 +571,14 @@
 
 lib/mc.ext
 
-vfs/extfs/a
-vfs/extfs/apt
 vfs/extfs/audio
-vfs/extfs/deb
-vfs/extfs/deba
-vfs/extfs/debd
-vfs/extfs/dpkg
 vfs/extfs/hp48
 vfs/extfs/lslR
-vfs/extfs/mailfs
-vfs/extfs/patchfs
-vfs/extfs/rpms
 vfs/extfs/uar
 vfs/extfs/uarj
 vfs/extfs/uha
 vfs/extfs/ulha 
 vfs/extfs/urar
-vfs/extfs/uzip
 vfs/extfs/uzoo
 
 doc/mc.1 doc/mcedit.1 doc/mcview.1 doc/mcserv.8
diff -urN mc-4.6.0-orig/vfs/extfs/Makefile.am mc-4.6.0/vfs/extfs/Makefile.am
--- mc-4.6.0-orig/vfs/extfs/Makefile.am	2004-02-06 21:06:20.000000000 +0100
+++ mc-4.6.0/vfs/extfs/Makefile.am	2004-02-06 21:11:58.898323552 +0100
@@ -8,46 +8,26 @@
 
 # Scripts that need adaptation to the local system - source files
 EXTFS_IN    = 			\
-	a.in			\
-	apt.in			\
 	audio.in		\
-	deb.in			\
-	deba.in			\
-	debd.in			\
-	dpkg.in			\
 	hp48.in			\
 	lslR.in			\
-	mailfs.in		\
-	patchfs.in		\
-	rpms.in			\
 	uar.in			\
 	uarj.in			\
 	uha.in			\
 	ulha.in			\
 	urar.in			\
-	uzip.in			\
 	uzoo.in
 
 # Scripts that need adaptation to the local system - files to install
 EXTFS_OUT = 			\
-	a			\
-	apt			\
 	audio			\
-	deb			\
-	deba			\
-	debd			\
-	dpkg			\
 	hp48			\
 	lslR			\
-	mailfs			\
-	patchfs			\
-	rpms			\
 	uar			\
 	uarj			\
 	uha			\
 	ulha			\
 	urar			\
-	uzip			\
 	uzoo
 
 if USE_VFS
diff -urN mc-4.6.0-orig/vfs/extfs/Makefile.in mc-4.6.0/vfs/extfs/Makefile.in
--- mc-4.6.0-orig/vfs/extfs/Makefile.in	2003-02-05 19:09:10.000000000 +0100
+++ mc-4.6.0/vfs/extfs/Makefile.in	2004-02-06 21:14:28.548573256 +0100
@@ -193,47 +193,26 @@
 
 # Scripts that need adaptation to the local system - source files
 EXTFS_IN = \
-	a.in			\
-	apt.in			\
 	audio.in		\
-	deb.in			\
-	deba.in			\
-	debd.in			\
-	dpkg.in			\
 	hp48.in			\
 	lslR.in			\
-	mailfs.in		\
-	patchfs.in		\
-	rpms.in			\
 	uar.in			\
 	uarj.in			\
 	uha.in			\
 	ulha.in			\
 	urar.in			\
-	uzip.in			\
 	uzoo.in
 
 
 # Scripts that need adaptation to the local system - files to install
 EXTFS_OUT = \
-	a			\
-	apt			\
 	audio			\
-	deb			\
-	deba			\
-	debd			\
-	dpkg			\
-	hp48			\
 	lslR			\
-	mailfs			\
-	patchfs			\
-	rpms			\
 	uar			\
 	uarj			\
 	uha			\
 	ulha			\
 	urar			\
-	uzip			\
 	uzoo
 
 
@@ -244,17 +223,14 @@
 subdir = vfs/extfs
 mkinstalldirs = $(SHELL) $(top_srcdir)/config/mkinstalldirs
 CONFIG_HEADER = $(top_builddir)/config.h
-CONFIG_CLEAN_FILES = a apt audio deb deba debd dpkg hp48 lslR mailfs \
-	patchfs rpms uar uarj uha ulha urar uzip uzoo
+CONFIG_CLEAN_FILES = audio hp48 lslR uar uarj uha ulha urar uzoo
 SCRIPTS = $(extfs_SCRIPTS)
 
 DIST_SOURCES =
 DATA = $(extfs_DATA)
 
-DIST_COMMON = README Makefile.am Makefile.in a.in apt.in audio.in \
-	deb.in deba.in debd.in dpkg.in hp48.in lslR.in mailfs.in \
-	patchfs.in rpms.in uar.in uarj.in uha.in ulha.in urar.in \
-	uzip.in uzoo.in
+DIST_COMMON = README Makefile.am Makefile.in audio.in hp48.in \
+	lslR.in uar.in uarj.in uha.in ulha.in urar.in uzoo.in
 all: all-am
 
 .SUFFIXES:
@@ -263,30 +239,12 @@
 	  $(AUTOMAKE) --gnu  vfs/extfs/Makefile
 Makefile: @MAINTAINER_MODE_TRUE@ $(srcdir)/Makefile.in  $(top_builddir)/config.status
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@ $(am__depfiles_maybe)
-a: $(top_builddir)/config.status a.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-apt: $(top_builddir)/config.status apt.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 audio: $(top_builddir)/config.status audio.in
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-deb: $(top_builddir)/config.status deb.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-deba: $(top_builddir)/config.status deba.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-debd: $(top_builddir)/config.status debd.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-dpkg: $(top_builddir)/config.status dpkg.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 hp48: $(top_builddir)/config.status hp48.in
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 lslR: $(top_builddir)/config.status lslR.in
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-mailfs: $(top_builddir)/config.status mailfs.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-patchfs: $(top_builddir)/config.status patchfs.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-rpms: $(top_builddir)/config.status rpms.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 uar: $(top_builddir)/config.status uar.in
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 uarj: $(top_builddir)/config.status uarj.in
@@ -297,8 +255,6 @@
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 urar: $(top_builddir)/config.status urar.in
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
-uzip: $(top_builddir)/config.status uzip.in
-	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 uzoo: $(top_builddir)/config.status uzoo.in
 	cd $(top_builddir) && $(SHELL) ./config.status $(subdir)/$@
 extfsSCRIPT_INSTALL = $(INSTALL_SCRIPT)
