diff -urN mc-4.6.0-orig/syntax/cvs-userlist.syntax mc-4.6.0/syntax/cvs-userlist.syntax
--- mc-4.6.0-orig/syntax/cvs-userlist.syntax	1970-01-01 01:00:00.000000000 +0100
+++ mc-4.6.0/syntax/cvs-userlist.syntax	2004-02-11 04:16:02.995824416 +0100
@@ -0,0 +1,20 @@
+wholechars abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._-
+
+context default
+	keyword : yellow black
+# people
+	keyword whole linestart ac-admin black white
+	keyword whole linestart ra-admin black white
+	keyword whole linestart nest-admin black white
+	keyword whole linestart cvsadmin black white
+	keyword whole linestart distfilesadmin black white
+# machines (hm... people vs machines... like in the Matrix ;))
+	keyword wholeleft linestart builder gray white
+	keyword wholeleft linestart buildlogs gray white
+	keyword wholeleft linestart cia gray white
+	keyword wholeleft linestart distfiles gray white
+
+context exclusive : : cyan
+	keyword , yellow black
+	
+
diff -urN mc-4.6.0-orig/syntax/Makefile.am mc-4.6.0/syntax/Makefile.am
--- mc-4.6.0-orig/syntax/Makefile.am	2004-02-11 03:21:32.000000000 +0100
+++ mc-4.6.0/syntax/Makefile.am	2004-02-11 04:17:12.818209800 +0100
@@ -5,6 +5,7 @@
 	ada95.syntax		\
 	c.syntax		\
 	changelog.syntax	\
+	cvs-userlist.syntax	\
 	diff.syntax		\
 	dos.syntax		\
 	fortran.syntax		\
@@ -24,6 +25,7 @@
 	perl.syntax		\
 	php.syntax		\
 	po.syntax		\
+	pld-update-todo.syntax	\
 	python.syntax		\
 	sh.syntax		\
 	smalltalk.syntax	\
diff -urN mc-4.6.0-orig/syntax/Makefile.in mc-4.6.0/syntax/Makefile.in
--- mc-4.6.0-orig/syntax/Makefile.in	2003-02-05 19:09:09.000000000 +0100
+++ mc-4.6.0/syntax/Makefile.in	2004-02-11 04:18:08.730709808 +0100
@@ -190,6 +190,7 @@
 	ada95.syntax		\
 	c.syntax		\
 	changelog.syntax	\
+	cvs-userlist.syntax	\
 	diff.syntax		\
 	dos.syntax		\
 	fortran.syntax		\
@@ -208,6 +209,7 @@
 	pascal.syntax		\
 	perl.syntax		\
 	php.syntax		\
+	pld-update-todo.syntax	\
 	po.syntax		\
 	python.syntax		\
 	sh.syntax		\
diff -urN mc-4.6.0-orig/syntax/pld-update-todo.syntax mc-4.6.0/syntax/pld-update-todo.syntax
--- mc-4.6.0-orig/syntax/pld-update-todo.syntax	1970-01-01 01:00:00.000000000 +0100
+++ mc-4.6.0/syntax/pld-update-todo.syntax	2004-02-11 04:16:02.996824264 +0100
@@ -0,0 +1,5 @@
+context default
+	keyword wholeleft	[OLD]\s*\s	brightred
+	keyword wholeleft	[NEW]\s*\n	green
+	keyword wholeright	(*)	white black
+
diff -urN mc-4.6.0-orig/syntax/Syntax mc-4.6.0/syntax/Syntax
--- mc-4.6.0-orig/syntax/Syntax	2004-02-11 03:21:32.000000000 +0100
+++ mc-4.6.0/syntax/Syntax	2004-02-11 04:16:02.997824112 +0100
@@ -125,6 +125,12 @@
 file Don_t_match_me Mail\sfolder ^From\s
 include mail.syntax
 
+file PLD-update-TODO PLD\sTODO
+include pld-update-todo.syntax
+
+file users CVS\suser\slist
+include cvs-userlist.syntax
+
 file .\* unknown
 include unknown.syntax
 
