diff -Nru mc-4.5.44/vfs/extfs/rpm mc-4.5.44.new/vfs/extfs/rpm
--- mc-4.5.44/vfs/extfs/rpm	Fri Apr  7 05:58:19 2000
+++ mc-4.5.44.new/vfs/extfs/rpm	Fri Apr  7 05:55:44 2000
@@ -99,6 +99,8 @@
 	echo "$FILEPREF 0 $DATE INFO/REQUIRENAME"
     rpm -qp --qf "[%{OBSOLETES}\n]" "$1" | grep "(none)" > /dev/null ||
 	echo "$FILEPREF 0 $DATE INFO/OBSOLETES"
+    rpm -qp --qf "[%{CONFLICTS}\n]" "$1" | grep "(none)" > /dev/null ||
+	echo "$FILEPREF 0 $DATE INFO/CONFLICTS"
     rpm -qp --qf "[%{PROVIDES}\n]" "$1" | grep "(none)" > /dev/null ||
 	echo "$FILEPREF 0 $DATE INFO/PROVIDES"
     test "`rpm -qp --qf \"%{CHANGELOGTEXT}\" \"$1\"`" = "(none)" ||
@@ -106,11 +108,12 @@
     else 
 	echo "$FILEPREF 0 $DATE INFO/REQUIRENAME"
 	echo "$FILEPREF 0 $DATE INFO/OBSOLETES"
+	echo "$FILEPREF 0 $DATE INFO/CONFLICTS"
 	echo "$FILEPREF 0 $DATE INFO/PROVIDES"
         echo "$FILEPREF 0 $DATE INFO/CHANGELOG"
     fi
 
-#    rpm -qlvp "$1" | sed -e 's/^\(..........\)[-t]* /\1 1 /'
+    rpm -qlvp "$1" | sed -e 's/^\(..........\)[-t]* /\1 1 /'
     echo "$FILEPREF 0 $DATE CONTENTS.cpio"
 }
 
@@ -135,6 +138,8 @@
 	INFO/COPYRIGHT)		rpm -qp --qf "%{COPYRIGHT}\n" "$1" >"$3"; exit 0;;
 	INFO/RPMVERSION)	rpm -qp --qf "%{RPMVERSION}\n" "$1" >"$3"; exit 0;;
 	INFO/REQUIRENAME)	rpm -qp --qf "[%{REQUIRENAME} %{REQUIREFLAGS:depflags} %{REQUIREVERSION}\n]" "$1" >"$3"; exit 0;;
+	INFO/OBSOLETES)		rpm -qp --qf "[%{OBSOLETES}\n]" "$1" >"$3"; exit 0;;
+	INFO/CONFLICTS)		rpm -qp --qf "[%{CONFLICTS}\n]" "$1" >"$3"; exit 0;;
 	INFO/PROVIDES)		rpm -qp --qf "[%{PROVIDES}\n]" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/PREIN)	rpm -qp --qf "%{RPMTAG_PREIN}\n" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/POSTIN)	rpm -qp --qf "%{RPMTAG_POSTIN}\n" "$1" >"$3"; exit 0;;
@@ -142,11 +147,21 @@
 	INFO/SCRIPTS/POSTUN)	rpm -qp --qf "%{RPMTAG_POSTUN}\n" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/VERIFYSCRIPT)	rpm -qp --qf "%{VERIFYSCRIPT}\n" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/ALL)		rpm -qp --scripts "$1" > "$3"; exit 0;;
+	INFO/SCRIPTS/TRIGGERSCRIPTS)	rpm -qp --triggers "$1" > "$3"; exit 0;;
 	INFO/SUMMARY)		rpm -qp --qf "%{SUMMARY}\n" "$1" > "$3"; exit 0;;
 	INFO/OS)		rpm -qp --qf "%{OS}\n" "$1" > "$3"; exit 0;;
 	INFO/CHANGELOG)		rpm -qp --qf "[* %{CHANGELOGTIME:date} %{CHANGELOGNAME}\n%{CHANGELOGTEXT}\n\n]\n" "$1" > "$3"; exit 0;;
 	INFO/SIZE)		rpm -qp --qf "%{SIZE} bytes\n" "$1" > "$3"; exit 0;;
 	CONTENTS.cpio)		rpm2cpio "$1" > "$3"; exit 0;;
+        *)
+		TMP_DIR=${TMPDIR:-/tmp}/mctmpdir.$$
+		mkdir $TMP_DIR || exit 1
+		cd $TMP_DIR
+		rpm2cpio "$1" | cpio -iumd --quiet "$2" >/dev/null
+		mv "$2" "$3"
+		cd /
+		rm -rf $TMP_DIR
+		;;
     esac
 }
 
