diff -Nru mc-4.5.40/vfs/extfs/rpm mc-4.5.40.new/vfs/extfs/rpm
--- mc-4.5.40/vfs/extfs/rpm	Wed Sep 29 22:39:34 1999
+++ mc-4.5.40.new/vfs/extfs/rpm	Tue Nov  2 02:02:02 1999
@@ -93,6 +93,8 @@
 	echo "$FILEPREF 0 $DATE INFO/REQUIRENAME"
     rpm -qp --qf "[%{OBSOLETES}\n]" "$1" | grep "(none)" > /dev/null ||
 	echo "$FILEPREF 0 $DATE INFO/OBSOLETES"
+    rpm -qp --qf "[%{CONFLICTS}\n]" "$1" | grep "(none)" > /dev/null ||
+	echo "$FILEPREF 0 $DATE INFO/CONFLICTS"
     rpm -qp --qf "[%{PROVIDES}\n]" "$1" | grep "(none)" > /dev/null ||
 	echo "$FILEPREF 0 $DATE INFO/PROVIDES"
     test "`rpm -qp --qf \"%{CHANGELOGTEXT}\" \"$1\"`" = "(none)" ||
@@ -100,6 +102,7 @@
     else 
 	echo "$FILEPREF 0 $DATE INFO/REQUIRENAME"
 	echo "$FILEPREF 0 $DATE INFO/OBSOLETES"
+	echo "$FILEPREF 0 $DATE INFO/CONFLICTS"
 	echo "$FILEPREF 0 $DATE INFO/PROVIDES"
         echo "$FILEPREF 0 $DATE INFO/CHANGELOG"
     fi
@@ -128,6 +131,8 @@
 	INFO/COPYRIGHT)		rpm -qp --qf "%{COPYRIGHT}\n" "$1" >"$3"; exit 0;;
 	INFO/RPMVERSION)	rpm -qp --qf "%{RPMVERSION}\n" "$1" >"$3"; exit 0;;
 	INFO/REQUIRENAME)	rpm -qp --qf "[%{REQUIRENAME} %{REQUIREFLAGS:depflags} %{REQUIREVERSION}\n]" "$1" >"$3"; exit 0;;
+	INFO/OBSOLETES)		rpm -qp --qf "[%{OBSOLETES}\n]" "$1" >"$3"; exit 0;;
+	INFO/CONFLICTS)		rpm -qp --qf "[%{CONFLICTS}\n]" "$1" >"$3"; exit 0;;
 	INFO/PROVIDES)		rpm -qp --qf "[%{PROVIDES}\n]" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/PREIN)	rpm -qp --qf "%{RPMTAG_PREIN}\n" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/POSTIN)	rpm -qp --qf "%{RPMTAG_POSTIN}\n" "$1" >"$3"; exit 0;;
