diff -uNr mc-4.5.54.orig/vfs/extfs/rpm mc-4.5.54/vfs/extfs/rpm
--- mc-4.5.54.orig/vfs/extfs/rpm	Mon May 14 10:56:08 2001
+++ mc-4.5.54/vfs/extfs/rpm	Mon May 14 10:38:34 2001
@@ -35,7 +35,11 @@
     echo "-r-xr-xr-x   1 root     root    39 $DATE INSTALL"
     echo "-r-xr-xr-x   1 root     root    39 $DATE UPGRADE"
     echo "dr-xr-xr-x   3 root     root	   0 $DATE INFO"
-    echo "$FILEPREF 0 $DATE INFO/NAME-VERSION-RELEASE"
+    if [ "`rpm -qp --qf \"%{SERIAL}\" \"$1\"`" = "(none)" ]; then
+	echo "$FILEPREF 0 $DATE INFO/NAME-VERSION-RELEASE"
+    else
+	echo "$FILEPREF 0 $DATE INFO/NAME-SERIAL:VERSION-RELEASE"
+    fi
     echo "$FILEPREF 0 $DATE INFO/GROUP"
     echo "$FILEPREF 0 $DATE INFO/BUILDHOST"
     echo "$FILEPREF 0 $DATE INFO/SOURCERPM"
@@ -102,11 +106,14 @@
 	echo "$FILEPREF 0 $DATE INFO/REQUIRENAME"
     rpm -qp --qf "[%{OBSOLETES}\n]" "$1" | grep "(none)" > /dev/null ||
 	echo "$FILEPREF 0 $DATE INFO/OBSOLETES"
+    rpm -qp --qf "[%{CONFLICTS}\n]" "$1" | grep "(none)" > /dev/null ||
+	echo "$FILEPREF 0 $DATE INFO/CONFLICTS"
     rpm -qp --qf "[%{PROVIDES}\n]" "$1" | grep "(none)" > /dev/null ||
 	echo "$FILEPREF 0 $DATE INFO/PROVIDES"
     test "`rpm -qp --qf \"%{CHANGELOGTEXT}\" \"$1\"`" = "(none)" ||
        echo "$FILEPREF 0 $DATE INFO/CHANGELOG"
     else 
+	echo "$FILEPREF 0 $DATE INFO/CONFLICTS"
 	echo "$FILEPREF 0 $DATE INFO/REQUIRENAME"
 	echo "$FILEPREF 0 $DATE INFO/OBSOLETES"
 	echo "$FILEPREF 0 $DATE INFO/PROVIDES"
@@ -124,7 +131,10 @@
 	INSTALL) echo "# Run this to install this RPM package" > "$3"; exit 0;;
 	UPGRADE) echo "# Run this to upgrade this RPM package" > "$3"; exit 0;;
 	ERROR) rpm -qip "$1" > /dev/null 2> "$3"; exit 0;;
-	INFO/NAME-VERSION-RELEASE)	rpm -qp --qf "%{NAME}-%{VERSION}-%{RELEASE}\n" "$1" > "$3"; exit 0;;
+	INFO/NAME-VERSION-RELEASE)
+		rpm -qp --qf "%{NAME}-%{VERSION}-%{RELEASE}\n" "$1" > "$3"; exit 0;;
+	INFO/NAME-SERIAL:VERSION-RELEASE)
+		rpm -qp --qf "%{NAME}-%{SERIAL}:%{VERSION}-%{RELEASE}\n" "$1" > "$3"; exit 0;;
 	INFO/RELEASE)		rpm -qp --qf "%{RELEASE}\n" "$1" > "$3"; exit 0;;
 	INFO/GROUP)		rpm -qp --qf "%{GROUP}\n" "$1" > "$3"; exit 0;;
 	INFO/DISTRIBUTION) 	rpm -qp --qf "%{DISTRIBUTION}\n" "$1" > "$3"; exit 0;;
@@ -139,27 +149,32 @@
 	INFO/COPYRIGHT)		rpm -qp --qf "%{COPYRIGHT}\n" "$1" >"$3"; exit 0;;
 	INFO/RPMVERSION)	rpm -qp --qf "%{RPMVERSION}\n" "$1" >"$3"; exit 0;;
 	INFO/REQUIRENAME)	rpm -qp --qf "[%{REQUIRENAME} %{REQUIREFLAGS:depflags} %{REQUIREVERSION}\n]" "$1" >"$3"; exit 0;;
+	INFO/OBSOLETES)		rpm -qp --qf "[%{OBSOLETES}\n]" "$1" >"$3"; exit 0;;
+	INFO/CONFLICTS)		rpm -qp --qf "[%{CONFLICTS}\n]" "$1" >"$3"; exit 0;;
 	INFO/PROVIDES)		rpm -qp --qf "[%{PROVIDES}\n]" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/PREIN)	rpm -qp --qf "%{RPMTAG_PREIN}\n" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/POSTIN)	rpm -qp --qf "%{RPMTAG_POSTIN}\n" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/PREUN)	rpm -qp --qf "%{RPMTAG_PREUN}\n" "$1" >"$3"; exit 0;;
 	INFO/SCRIPTS/POSTUN)	rpm -qp --qf "%{RPMTAG_POSTUN}\n" "$1" >"$3"; exit 0;;
-	INFO/SCRIPTS/VERIFYSCRIPT)	rpm -qp --qf "%{VERIFYSCRIPT}\n" "$1" >"$3"; exit 0;;
+	INFO/SCRIPTS/VERIFYSCRIPT)
+		rpm -qp --qf "%{VERIFYSCRIPT}\n" "$1" >"$3"; exit 0;;
+	INFO/SCRIPTS/TRIGGERSCRIPTS)
+		rpm -qp --triggers "$1" > "$3"; exit 0;;
 	INFO/SCRIPTS/ALL)		rpm -qp --scripts "$1" > "$3"; exit 0;;
 	INFO/SUMMARY)		rpm -qp --qf "%{SUMMARY}\n" "$1" > "$3"; exit 0;;
 	INFO/OS)		rpm -qp --qf "%{OS}\n" "$1" > "$3"; exit 0;;
 	INFO/CHANGELOG)		rpm -qp --qf "[* %{CHANGELOGTIME:date} %{CHANGELOGNAME}\n%{CHANGELOGTEXT}\n\n]\n" "$1" > "$3"; exit 0;;
 	INFO/SIZE)		rpm -qp --qf "%{SIZE} bytes\n" "$1" > "$3"; exit 0;;
+	CONTENTS.cpio)		rpm2cpio "$1" > "$3"; exit 0;;
 	*)
-	    TMPDIR=/tmp/mctmpdir.$$
+	    TMPDIR=${TMPDIR:-/tmp}/mctmpdir.$$
 	    mkdir $TMPDIR || exit 1
 	    cd $TMPDIR
-	    rpm2cpio "$1" | cpio -iumd --quiet "$2" >/dev/null
+	    rpm2cpio "$1" | cpio -iumd --quiet "$2" "./$2" >/dev/null
 	    mv "$2" "$3"
 	    cd /
 	    rm -rf $TMPDIR;;
 
-	CONTENTS.cpio)		rpm2cpio "$1" > "$3"; exit 0;;
     esac
 }
 
