diff -urN mc-4.5.52.org/vfs/samba/Makefile.in mc-4.5.52/vfs/samba/Makefile.in
--- mc-4.5.52.org/vfs/samba/Makefile.in	Thu Mar 15 03:05:53 2001
+++ mc-4.5.52/vfs/samba/Makefile.in	Sun Mar 25 17:35:08 2001
@@ -51,7 +51,7 @@
 LOCKDIR = @lockdir@
 
 # The directory where code page definition files go
-CODEPAGEDIR = $(LIBDIR)/codepages
+CODEPAGEDIR = /etc/samba/codepages
 
 # The current codepage definition list.
 CODEPAGELIST= 437 737 850 852 861 932 866 949 950 936
diff -urN mc-4.5.52.org/vfs/smbfs.c mc-4.5.52/vfs/smbfs.c
--- mc-4.5.52.org/vfs/smbfs.c	Thu Mar 15 21:16:28 2001
+++ mc-4.5.52/vfs/smbfs.c	Sun Mar 25 17:30:35 2001
@@ -375,7 +375,7 @@
 static int
 smbfs_init(vfs *me)
 {
-	char *servicesf = "/etc/smb.conf";
+	char *servicesf = "/etc/samba/smb.conf";
 
 /*	DEBUGLEVEL = 4;	*/
 
