--- mc-4.6.0/src/Makefile.am.orig	Tue Oct 15 17:33:04 2002
+++ mc-4.6.0/src/Makefile.am	Sun Feb 16 12:51:16 2003
@@ -7,9 +7,9 @@
 if CONS_SAVER
 pkglib_PROGRAMS = cons.saver
 AM_CPPFLAGS = -DDATADIR=\""$(pkgdatadir)/"\" -DLOCALEDIR=\""$(localedir)"\" \
-	-DLIBDIR=\""$(pkglibdir)"\"
+	-DLIBDIR=\""$(pkglibdir)"\" -I/usr/include/popt
 else
-AM_CPPFLAGS = -DDATADIR=\""$(pkgdatadir)/"\" -DLOCALEDIR=\""$(localedir)"\"
+AM_CPPFLAGS = -DDATADIR=\""$(pkgdatadir)/"\" -DLOCALEDIR=\""$(localedir)"\" -I/usr/include/popt
 endif
 
 noinst_PROGRAMS = man2hlp
@@ -38,7 +38,7 @@
 endif
 
 mc_LDADD = $(EDITLIB) $(SLANGLIB) $(VFSLIB) \
-	$(INTLLIBS) $(GLIB_LIBS) $(MCLIBS) $(LIBICONV)
+	$(INTLLIBS) $(GLIB_LIBS) $(MCLIBS) $(LIBICONV) -lpopt
 
 CHARSET_SRC = charsets.c charsets.h selcodepage.c selcodepage.h
 
@@ -53,8 +53,8 @@
 	keyxdef.c layout.c layout.h learn.c learn.h listmode.c		\
 	listmode.h main.c main.h menu.c menu.h				\
 	mountlist.c mountlist.h mouse.c mouse.h myslang.h option.c	\
-	option.h panel.h panelize.c panelize.h poptalloca.h popt.c	\
-	poptconfig.c popt.h popthelp.c poptint.h poptparse.c profile.c	\
+	option.h panel.h panelize.c panelize.h \
+	profile.c	\
 	profile.h regex.c rxvt.c screen.c setup.c setup.h slint.c	\
 	subshell.c subshell.h terms.c text.c textconf.c textconf.h	\
 	tree.c tree.h treestore.c treestore.h tty.h user.c user.h	\
