--- mc-4.5.51/src/background.c.stderr	Mon Jul  3 15:32:10 2000
+++ mc-4.5.51/src/background.c	Fri Aug  4 13:38:01 2000
@@ -146,7 +146,7 @@
 	close (1);
 	close (2);
 
-	if ((nullfd = open ("/dev/null", O_RDONLY)) != -1){
+	if ((nullfd = open ("/dev/null", O_RDWR)) != -1){
 	    while (dup2 (nullfd, 0) == -1 && errno == EINTR)
 		    ;
 	    while (dup2 (nullfd, 1) == -1 && errno == EINTR)
--- mc-4.5.51/gnome/gutil.c.stderr	Fri Aug  4 13:47:52 2000
+++ mc-4.5.51/gnome/gutil.c	Fri Aug  4 13:48:13 2000
@@ -159,10 +159,10 @@
 		open ("/dev/null", O_APPEND);
 
 		/* stdout */
-		open ("/dev/null", O_RDONLY);
+		open ("/dev/null", O_WRONLY);
 
 		/* stderr */
-		open ("/dev/null", O_RDONLY);
+		open ("/dev/null", O_WRONLY);
 		
 		if (!(flags & EXECUTE_WAIT))
 			*pid = fork ();
