--- doc/mc.1.in
+++ doc/mc.1.in
@@ -1385,8 +1385,9 @@
 	od -c %f
 
 B	Edit a bug report and send it to root
-	vi /tmp/mail.$$
-	mail -s "Midnight Commander bug" root < /tmp/mail.$$
+	I=`mktemp /tmp/mailXXXXXX`; export I
+	vi $I
+	mail -s "Midnight Commander bug" root < $I
 
 M	Read mail
 	emacs -f rmail
--- lib/mc.menu
+++ lib/mc.menu
@@ -14,9 +14,10 @@
 	
 
 0       Edit a bug report and send it to root
-        ${EDITOR-vi} /tmp/mail.$$
-	test -r /tmp/mail.$$ && mail root < /tmp/mail.$$
-	rm -f /tmp/mail.$$
+        I=`mktemp /tmp/mail.XXXXXX`; export I
+        ${EDITOR-vi} $I
+        test -r $I && mail root < $I
+        rm -f $I
 
 =+ f \.1$ | f \.3$ | f \.4$ | f \.5$ | f \.6$ | f \.7$ | f \.8$ | f \.man$ & t r
 1       Display the file with roff -man
@@ -112,8 +113,8 @@
 	CHECK=`awk '{print $1 ; exit}' %f` 2>/dev/null
 	case $CHECK in
 	  Newsgroups:|Path:)
-	      cp %f /tmp/%f.$$ && sed '/^'"$CHECK"' /,/^$/d' /tmp/%f.$$ > %f
-              [ "$?" = "0" ] && rm /tmp/%f.$$
+	      I=`mktemp /tmp/news.XXXXXX`; export I;cp %f $I && sed '/^'"$CHECK"' /,/^$/d' $I > %f
+	      [ "$?" = "0" ] && rm $I
 	      echo %f: header removed
 		;;
 	  *)
@@ -126,7 +127,7 @@
 	set %u
 	while [ -n "$1" ]; do
 	  CHECK=`sed 1q < $1|awk '{print $1}'` 2>/dev/null
-	  WFILE=/tmp/${1}.$$
+	  WFILE=`mktemp /tmp/news.XXXXXX`; export WFILE;
 	  case $CHECK in
 	    Newsgroups:|Path:)
 	      cp $1 $WFILE && sed '/^'"$CHECK"' /,/^$/d' $WFILE > $1
