--- 1/src/ChangeLog
+++ 1/src/ChangeLog
@@ -1 +1,6 @@
+2002-07-03  Pavel Roskin  <proski@gnu.org>
+
+       * subshell.c (sigchld_handler): Remove subshell_pty from the
+       channel list if the subshell has died.
+
 2002-07-03  Andrew V. Samoilov  <kai@cmail.ru>

--- 1/src/subshell.c.orig	Sat Jul 20 15:20:10 2002
+++ 1/src/subshell.c	Sat Jul 20 15:21:50 2002
@@ -939,6 +939,7 @@
 #  endif
 	{
 	    subshell_alive = FALSE;
+	    delete_select_channel (subshell_pty);
 	    if (WIFEXITED (status) && WEXITSTATUS (status) != FORK_FAILURE)
 		quit |= SUBSHELL_EXIT;  /* Exited normally */
 	}
