--- mc-4.5.53/configure.in	Sat Apr 14 02:40:35 2001
+++ mc-4.5.53.new/configure.in~	Fri Apr 13 03:05:02 2001
@@ -50,7 +50,7 @@
 
 ALL_LINGUAS="az ca cs da de el es es_ES fi fr hu it ja ko lv nl no pl pt_BR ro ru sl sv sk uk ta tr wa zh_TW.Big5 zh_CN.GB2312"
 
-AM_GNOME_GETTEXT
+AM_GNU_GETTEXT
 
 dnl
 dnl Enforce coding standards
--- mc-4.5.53/src/Makefile.in~	Sat Apr 14 02:42:51 2001
+++ mc-4.5.53/src/Makefile.in	Sun Apr 22 07:37:06 2001
@@ -14,7 +14,7 @@
 LIBS     = $(XLIBS) @TERMNET@ $(XLIB) @TERMNET@ @MCLIBS@ -lpopt
 OURLIBS  = @LVFS@ @LSLANG@ @LEDIT@ @LGPM@ @LINTL@
 DEPLIBS  = $(top_builddir)/vfs/@LIBVFS@ $(top_builddir)/slang/@LIBSLANG@ \
-	$(top_builddir)/edit/@LIBEDIT_A@ @INTLDEPS@
+	$(top_builddir)/edit/@LIBEDIT_A@
 INSTALL  = @INSTALL@
 
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
--- mc-4.5.53/gnome/Makefile.in~	Sat Apr 14 02:42:52 2001
+++ mc-4.5.53/gnome/Makefile.in	Sun Apr 22 07:38:19 2001
@@ -17,7 +17,7 @@
 LIBS = $(XLIBS) @TERMNET@
 EXTRALIBS = -L../vfs @LVFS@ -L../gtkedit -lgtkedit @LINTL@
 DEPLIBS = $(top_builddir)/vfs/@LIBVFS@ \
-	$(top_builddir)/gtkedit/@libgtkedit@ @INTLDEPS@
+	$(top_builddir)/gtkedit/@libgtkedit@
 INSTALL = @INSTALL@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_DATA = @INSTALL_DATA@
