diff -urN mc-4.5.52.org/src/dir.c mc-4.5.52/src/dir.c
--- mc-4.5.52.org/src/dir.c	Mon Mar 19 20:44:47 2001
+++ mc-4.5.52/src/dir.c	Mon Mar 19 21:04:25 2001
@@ -82,7 +82,7 @@
  * Ideally, the user should be able to select between all three functions -
  * strcoll() (if available), strcmp() and g_strcasecmp().
  */
-#ifdef HAVE_STRCOLL
+#if defined(HAVE_STRCOLL) && 0
 #define string_sortcomp(a,b) (case_sensitive ? strcoll (a,b) : g_strcasecmp (a,b))
 #else
 #define string_sortcomp(a,b) (case_sensitive ? strcmp (a,b) : g_strcasecmp (a,b))
