--- mc-4.5.55/configure.in.orig	Fri Jan 31 11:57:04 2003
+++ mc-4.5.55/configure.in	Fri Jan 31 11:59:58 2003
@@ -39,7 +39,7 @@
     X11_WWW=lynx
 fi
 
-ALL_LINGUAS="az ca cs da de el es es_ES fi fr hu it ja ko lv nl no pl pt_BR ro ru sk sl sv uk ta tr wa zh_TW.Big5 zh_CN.GB2312"
+ALL_LINGUAS="az ca cs da de el es es_ES fi fr hu it ja ko lv nl no pl pt_BR ro ru sk sl sv uk ta tr wa zh_TW zh_CN"
 
 dnl
 dnl Internationalization
