diff -urNp mc-4.7.0-pre4.orig/src/main.c mc-4.7.0-pre4/src/main.c
--- mc-4.7.0-pre4.orig/src/main.c	2009-10-30 19:44:39.000000000 +0100
+++ mc-4.7.0-pre4/src/main.c	2009-11-07 17:14:39.000000000 +0100
@@ -1403,6 +1403,9 @@ init_xterm_support (void)
 	|| strcmp (termvalue, "Eterm") == 0
 	|| strcmp (termvalue, "dtterm") == 0) {
 	xterm_flag = 1;
+#ifdef HAVE_SLANG
+	tty_display_8bit (0);
+#endif
 
 	/* Default to the standard xterm sequence */
 	if (!xmouse_seq) {
